diff --git a/src/core/runtime/detail/runtime.cc b/src/core/runtime/detail/runtime.cc
index 04cb5b15..3bc2556a 100644
--- a/src/core/runtime/detail/runtime.cc
+++ b/src/core/runtime/detail/runtime.cc
@@ -1797,7 +1797,7 @@ RuntimeManager::State RuntimeManager::state() const noexcept { return state_; }
 
 void RuntimeManager::reset() noexcept
 {
-  rt_.reset();
+  //rt_.reset();
   state_ = State::FINALIZED;
 }
 
