diff --git a/MaxText/configs/base.yml b/MaxText/configs/base.yml
index 4c4a6751..6175cc1e 100644
--- a/MaxText/configs/base.yml
+++ b/MaxText/configs/base.yml
@@ -330,7 +330,7 @@ data_shuffle_seed: 0
 init_weights_seed: 0
 
 # You may disable clipping by setting gradient_clipping_threshold to zero.
-gradient_clipping_threshold: 1.0
+gradient_clipping_threshold: 0.0
 
 # Instead of updating the weights every step, you may effectively use a larger
 # batch by accumulating the gradient over a set of steps.
diff --git a/MaxText/configs/models/gpt3-24layers.yml b/MaxText/configs/models/gpt3-24layers.yml
new file mode 100644
index 00000000..8a8f5434
--- /dev/null
+++ b/MaxText/configs/models/gpt3-24layers.yml
@@ -0,0 +1,43 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing software
+# distributed under the License is distributed on an "AS IS" BASIS
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# model config for gpt3-175b
+
+base_emb_dim: 12288
+base_num_query_heads: 96
+base_num_kv_heads: 96
+base_mlp_dim: 49152
+base_num_decoder_layers: 24
+head_dim: 128
+trainable_position_size: 16384
+mlp_activations: ["gelu"]
+vocab_size: 50304
+enable_dropout: False
+logits_via_embedding: True
+normalize_embedding_logits: False
+logits_dot_in_fp32: False
+normalization_layer_epsilon: 1.e-05
+use_iota_embed: True
+fused_qkv: True
+opt_type: "adam_pax"
+decoder_block: "gpt3"
+dataset_type: "synthetic"
+gradient_clipping_threshold: 0.
+adam_b1: 0.9
+adam_b2: 0.95
+adam_eps: 1.e-8
+adam_weight_decay: 0.1
+checkpoint_period: 10_000
+target_eval_loss: 2.69
+eval_per_device_batch_size: 1
diff --git a/MaxText/configs/models/gpt3-xl.yml b/MaxText/configs/models/gpt3-xl.yml
new file mode 100644
index 00000000..fa0946c1
--- /dev/null
+++ b/MaxText/configs/models/gpt3-xl.yml
@@ -0,0 +1,55 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing software
+# distributed under the License is distributed on an "AS IS" BASIS
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# model config for gpt3-6b
+
+base_emb_dim: 2048
+base_num_query_heads: 24
+base_num_kv_heads: 24
+base_mlp_dim: 2048
+base_num_decoder_layers: 24
+head_dim: 128
+max_target_length: 1024
+trainable_position_size: 16384
+mlp_activations: ["gelu"]
+vocab_size: 32768
+enable_dropout: False
+logits_via_embedding: True
+normalize_embedding_logits: False
+logits_dot_in_fp32: False
+normalization_layer_epsilon: 1.e-05
+use_iota_embed: True
+fused_qkv: True
+opt_type: "adam_pax"
+decoder_block: "gpt3"
+#dataset_path: "gs://mlperf-llm-public2"
+#dataset_name: "c4/en:3.0.4"
+#eval_dataset_name: "c4/en:3.0.5"
+dataset_type: hf
+hf_path: '/home/jwilke/Programs/workspaces/realm-jax-refactor/workflows-update/maxtext/wikitext'
+hf_data_dir: 'wikitext-103-v1'
+hf_train_files: '/home/jwilke/Programs/workspaces/realm-jax-refactor/workflows-update/maxtext/wikitext/wikitext-103-v1/train*.parquet'
+# set eval_interval > 0 to use the specified eval dataset, otherwise, only metrics on the train set will be calculated.
+eval_interval: 10000
+hf_eval_split: 'validation'
+hf_eval_files: ''
+# for HF pipeline, tokenizer_path can be a path in HuggingFace Hub,
+# or a local path containing tokenizer in a format supported by transformers.AutoTokenizer
+tokenizer_path: 'google-t5/t5-large'  # for using https://huggingface.co/google-t5/t5-large
+hf_access_token: ''  # provide token if using gated dataset or tokenizer
+gradient_clipping_threshold: 1.
+adam_b1: 0.9
+adam_b2: 0.95
+adam_eps: 1.e-8
+adam_weight_decay: 0.1
diff --git a/MaxText/input_pipeline/_hf_data_processing.py b/MaxText/input_pipeline/_hf_data_processing.py
index 76ef7ce0..a1ed3f34 100644
--- a/MaxText/input_pipeline/_hf_data_processing.py
+++ b/MaxText/input_pipeline/_hf_data_processing.py
@@ -117,7 +117,7 @@ def preprocessing_pipeline(
       data_source=dataset,
       operations=operations,
       sampler=dummy_index_sampler,
-      worker_count=1,  # only supports one worker for now, more workers results in duplicated data
+      worker_count=0,  # only supports one worker for now, more workers results in duplicated data
       worker_buffer_size=1,
       read_options=grain.ReadOptions(num_threads=num_threads, prefetch_buffer_size=128),
   )
diff --git a/MaxText/max_utils.py b/MaxText/max_utils.py
index 897674e2..d07de3c0 100644
--- a/MaxText/max_utils.py
+++ b/MaxText/max_utils.py
@@ -29,6 +29,7 @@ import numpy as np
 import jax
 import jax.numpy as jnp
 from jax.experimental import mesh_utils
+from jax.experimental.pjit import pjit
 import orbax.checkpoint as ocp
 import orbax.checkpoint.experimental.emergency.checkpoint_manager as emergency_checkpoint_manager
 
@@ -95,8 +96,11 @@ def summarize_size_from_pytree(params):
 
 
 def initialize_summary_writer(config):
-  summary_writer_path = os.path.join(config.tensorboard_dir, config.run_name)
-  return writer.SummaryWriter(summary_writer_path) if jax.process_index() == 0 else None
+  return (
+      writer.SummaryWriter(config.tensorboard_dir)
+      if jax.process_index() == 0
+      else None
+  )
 
 
 def close_summary_writer(summary_writer):
@@ -128,14 +132,16 @@ def write_metrics_locally(metrics, step, config, file, is_training=True):
 
 def add_config_to_summary_writer(config, summary_writer):
   """Writes config params to tensorboard"""
-  if jax.process_index() == 0:
+  #if jax.process_index() == 0:
+  if False:
     for key, value in config.get_keys().items():
       add_text_to_summary_writer(key, str(value), summary_writer)
 
 
 def add_text_to_summary_writer(key, value, summary_writer):
   """Writes given key-value pair to tensorboard as text/summary"""
-  if jax.process_index() == 0:
+  #if jax.process_index() == 0:
+  if False:
     summary_writer.add_text(key, value)
 
 
@@ -163,6 +169,7 @@ def write_config_raw_keys_for_gcs(raw_keys):
   """Writes config raw keys to GCS"""
   if not raw_keys["save_config_to_gcs"] or jax.process_index() != 0:
     return
+  return
   max_logging.log("Writing config to GCS...")
 
   raw_keys_dict = dict(raw_keys)
@@ -225,7 +232,6 @@ def maybe_initialize_jax_distributed_system(raw_keys):
       initialize_jax_for_tpu_with_emergency_checkpointing(raw_keys)
     max_logging.log("Jax distributed system initialized!")
 
-
 def initialize_jax_for_gpu():
   """Jax distributed initialize for GPUs."""
   if os.environ.get("JAX_COORDINATOR_IP") is not None:
@@ -244,17 +250,17 @@ def initialize_jax_for_cpu():
   coordinator_ip_address = get_coordinator_ip_address()
   coordinator_address = coordinator_ip_address + ":1234"  # JAX coordinator port used in XPK
   # Env variables to be set in XPK or otherwise
-  job_index = int(os.environ.get("JOB_INDEX"))
-  job_completion_index = int(os.environ.get("JOB_COMPLETION_INDEX"))
-  processes_in_job = int(os.environ.get("PROCESSES_IN_JOB"))
-  pid = job_index * processes_in_job + job_completion_index
-  max_logging.log(f" Jax process id is {pid} ")
-  # Explicit initialize is needed only for CPUs
-  jax.distributed.initialize(
-      coordinator_address=coordinator_address,
-      process_id=pid,
-      num_processes=int(os.environ.get("JAX_PROCESS_COUNT")),
-  )
+  if (job_index := os.environ.get("JOB_INDEX")) is not None:
+    job_completion_index = int(os.environ.get("JOB_COMPLETION_INDEX"))
+    processes_in_job = int(os.environ.get("PROCESSES_IN_JOB"))
+    pid = job_index * processes_in_job + job_completion_index
+    max_logging.log(f" Jax process id is {pid} ")
+    # Explicit initialize is needed only for CPUs
+    jax.distributed.initialize(
+        coordinator_address=coordinator_address,
+        process_id=pid,
+        num_processes=int(os.environ.get("JAX_PROCESS_COUNT")),
+    )
 
 
 def initialize_jax_for_tpu_with_emergency_checkpointing(raw_keys):
@@ -498,6 +504,36 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
   return state, state_mesh_annotations
 
 
+def bind_mesh(pjitted_fn, global_mesh: jax.sharding.Mesh):
+  """Wraps a pjitted_fn with a mesh context."""
+
+  def call(*args):
+    with global_mesh:
+      return pjitted_fn(*args)
+
+  def lower(*args, **kwargs):
+    with global_mesh:
+      return pjitted_fn.lower(*args, **kwargs)
+
+  call.lower = lower
+  return call
+
+
+def initialize_state_with_sharding(init_state_partial, mesh, state_sharding, rng_sharding, rng):
+
+  with mesh:
+    jitted = pjit(
+        init_state_partial,
+        in_shardings=rng_sharding,
+        out_shardings=state_sharding,
+    )
+  jitted = bind_mesh(jitted, mesh)
+
+  state = jitted(rng)
+
+  state = unbox_logicallypartioned(state)
+  return state
+
 def setup_training_state(model, data_iterator, tx, config, rng, mesh, checkpoint_manager):
   is_training = True
   return setup_initial_state(
@@ -512,6 +548,21 @@ def setup_training_state(model, data_iterator, tx, config, rng, mesh, checkpoint
   )
 
 
+def get_abstract_state_no_sharding(model, tx, config, rng, mesh, is_training=True):
+  """Get a shaped abstraction of the state (including optimizer)"""
+  init_state_partial = functools.partial(init_initial_state, model, tx, config, is_training)
+
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
+    abstract_state = jax.eval_shape(init_state_partial, rng)
+
+  state_logical_annotations = nn.get_partition_spec(abstract_state)
+  with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
+    state_mesh_annotations = nn.logical_to_mesh(state_logical_annotations)
+
+  abstract_state = unbox_logicallypartioned(abstract_state)
+  return init_state_partial, abstract_state, state_mesh_annotations
+
+
 def setup_initial_state(
     model,
     data_iterator,
diff --git a/MaxText/maxtext_utils.py b/MaxText/maxtext_utils.py
index 7e93ad06..7b6516a0 100644
--- a/MaxText/maxtext_utils.py
+++ b/MaxText/maxtext_utils.py
@@ -19,9 +19,17 @@ limitations under the License.
 
 import jax
 import optax
+import optimizers
 import max_utils
 from jax.sharding import PartitionSpec as P
 from jax.experimental.serialize_executable import deserialize_and_load
+from flax.linen import partitioning as nn_partitioning
+from jax.experimental.pjit import pjit, AUTO
+
+try:
+  from legate.jax import MeshWrapper
+except ImportError:
+  pass
 
 
 import pickle
@@ -30,6 +38,108 @@ from input_pipeline import input_pipeline_interface
 
 OVERWRITE_WITH_GRADIENT = "_overwrite_with_gradient"
 
+def bind_mesh(pjitted_fn, global_mesh: jax.sharding.Mesh):
+  """Wraps a pjitted_fn with a mesh context."""
+
+  def call(*args):
+    with global_mesh:
+      return pjitted_fn(*args)
+
+  def lower(*args, **kwargs):
+    with global_mesh:
+      return pjitted_fn.lower(*args, **kwargs)
+
+  call.lower = lower
+  return call
+
+
+
+def jit_and_compile(
+    func,
+    func_input_args,
+    func_input_kwargs,
+    mesh,
+    in_shardings,
+    out_shardings,
+    static_argnums,
+    donate_argnums,
+):
+  """Jit, lower, and compile func."""
+  jitted = pjit(
+      func,
+      in_shardings=in_shardings,
+      out_shardings=out_shardings,
+      static_argnums=static_argnums,
+      donate_argnums=donate_argnums,
+  )
+
+  jitted = bind_mesh(jitted, mesh)
+
+  with MeshWrapper.lower_mode():
+    lowered = jitted.lower(*func_input_args, **func_input_kwargs)
+  with MeshWrapper.compile_mode():
+    compiled = lowered.compile()
+
+  sharding_str_arr = []
+  def check(inp, inp_sharding, out):
+    sharding_str_arr.append(f"{inp.shape} {inp.sharding} {out}")
+    if not inp_sharding.is_equivalent_to(out, len(inp.shape)):
+      raise Exception("\n".join(sharding_str_arr) + " mismatch")
+
+  jax.tree_map(check, func_input_args[0], compiled.input_shardings[0][0], compiled.output_shardings[0])
+
+  return compiled
+
+def get_and_compile_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config, shaped_train_args):
+  transformer_shape = [(size if (name != 'stage') else 1) for name, size in mesh.shape.items()]
+  mesh = MeshWrapper(mesh, transformer_shape)
+
+  """Get the shardings (both state and data) for train_step"""
+  functional_train = get_functional_train_step(train_step, model, config)
+  functional_train.__name__ = "train_step"
+  data_pspec = P(*config.data_sharding)
+
+  def canonicalize_sharding(p):
+    if isinstance(p, jax.sharding.NamedSharding) or isinstance(p, jax.sharding.GSPMDSharding):
+      return p
+    return AUTO(mesh)
+
+  state_mesh_shardings = jax.tree_util.tree_map(canonicalize_sharding, state_mesh_annotations)
+  data_sharding = jax.tree_util.tree_map(lambda p: jax.sharding.NamedSharding(mesh, p), data_pspec)
+  in_shardings = (state_mesh_shardings, data_sharding, None)  # State, batch, rng
+
+  out_shardings = (state_mesh_shardings, None)
+
+   # get shapes from config
+  #shaped_train_args, shaped_train_kwargs = get_shaped_inputs(mesh, model, config)
+  shaped_train_kwargs = {}
+
+  static_argnums = ()  # We partial out the static argnums of model and config
+  donate_argnums = 0  # This is the index of the state - we allow the compiler to make use of this memory.
+
+  from jax.lax import with_sharding_constraint
+  from legate.jax import autoshard
+  def autoshard_train_fn(train_state, *args):
+    sharded_train_state = jax.tree_map(lambda x, y: with_sharding_constraint(x, y), train_state, state_mesh_annotations)
+    return functional_train(sharded_train_state, *args)
+
+  with mesh, autoshard(True), nn_partitioning.axis_rules(config.logical_axis_rules):
+    compiled = jit_and_compile(
+        autoshard_train_fn,
+        shaped_train_args,
+        shaped_train_kwargs,
+        mesh,
+        in_shardings,
+        out_shardings,
+        static_argnums,
+        donate_argnums,
+    )
+
+  # extract in/out-sharding from compiled function
+  in_shardings = compiled.input_shardings[0]
+  out_shardings = compiled.output_shardings[0]
+
+  return compiled, in_shardings, out_shardings, static_argnums, donate_argnums
 
 def get_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config):
   """Get the shardings (both state and data) for train_step"""
diff --git a/MaxText/optimizers.py b/MaxText/optimizers.py
index 6d42644f..332a9590 100644
--- a/MaxText/optimizers.py
+++ b/MaxText/optimizers.py
@@ -45,6 +45,8 @@ def get_optimizer(config, learning_rate_schedule):
         epsilon_root=config.adam_eps_root,
         weight_decay=config.adam_weight_decay,
     )
+  elif config.opt_type == "sgd":
+    return optax.sgd(learning_rate_schedule)
   else:
     raise ValueError(f"{config.opt_type=} is not a supported.")
 
diff --git a/MaxText/pyconfig.py b/MaxText/pyconfig.py
index 2ff2926e..883809a8 100644
--- a/MaxText/pyconfig.py
+++ b/MaxText/pyconfig.py
@@ -163,10 +163,20 @@ def validate_model_name(s: str) -> bool:
       "gpt3-22b",
       "gpt3-6b",
       "gpt3-52k",
+      "gpt3-xl",
+      "gpt3-24layers",
   )
   if s not in valid_model_names:
     raise ValueError("Invalid model name was passed. Valid options ", valid_model_names)
 
+def validate_overrides_consistent(keys1: list[str], keys2: list[str], raw_keys: dict, raw_keys_old: dict):
+  overwritten_keys = [k for k in keys1 if k in keys2]
+  for key in overwritten_keys:
+      if raw_keys_old[key] != raw_keys[key]:
+          raise ValueError(
+              f"Key {key} inconsistently overwritten by argument({raw_keys_old[key]})"
+              f" and model({raw_keys[key]}). This isn't allowed."
+          )
 
 def validate_no_keys_overwritten_twice(keys1: list[str], keys2: list[str]):
   overwritten_keys = [k for k in keys1 if k in keys2]
@@ -276,9 +286,12 @@ class _HyperParameters:
     raw_keys = OrderedDict()
     keys_from_env_and_command_line = self._update_from_env_and_command_line(raw_keys, raw_data_from_yaml, argv, **kwargs)
     max_logging.log(f"Updating keys from env and command line: {keys_from_env_and_command_line}")
+    raw_keys_before_model = raw_keys.copy()
     keys_from_model = _HyperParameters.update_model_vars(argv[1], raw_keys, config_name)
     max_logging.log(f"Updating keys from model: {keys_from_model}")
-    validate_no_keys_overwritten_twice(keys_from_env_and_command_line, keys_from_model)
+
+    validate_overrides_consistent(keys_from_env_and_command_line, keys_from_model, raw_keys, raw_keys_before_model)
+    #validate_no_keys_overwritten_twice(keys_from_env_and_command_line, keys_from_model)
 
     # We initialize the jax distributed system here because it must be done before device backend is initialized.
     max_utils.maybe_initialize_jax_distributed_system(raw_keys)
@@ -343,7 +356,7 @@ class _HyperParameters:
     raw_keys["num_slices"] = get_num_slices(raw_keys)
     raw_keys["quantization_local_shard_count"] = get_quantization_local_shard_count(raw_keys)
 
-    if using_pipeline_parallelism(raw_keys):
+    if False and using_pipeline_parallelism(raw_keys):
       raw_keys["using_pipeline_parallelism"] = True
       num_stages = int(raw_keys["ici_pipeline_parallelism"] * raw_keys["dcn_pipeline_parallelism"])
       if raw_keys["num_pipeline_repeats"] == -1:
diff --git a/MaxText/train.py b/MaxText/train.py
index 182dd75e..a24f9ad2 100644
--- a/MaxText/train.py
+++ b/MaxText/train.py
@@ -26,7 +26,7 @@ import sys
 import functools
 import time
 
-from typing import Sequence, Optional
+from typing import Optional, Sequence
 from absl import app
 from flax import linen as nn
 from flax.linen import partitioning as nn_partitioning
@@ -47,6 +47,7 @@ from vertex_tensorboard import VertexTensorboardManager
 # Placeholder: internal
 
 from input_pipeline.input_pipeline_interface import create_data_iterator
+from input_pipeline import input_pipeline_interface
 from layers import models
 
 import jax.numpy as jnp
@@ -63,6 +64,7 @@ from layers import quantizations
 
 from ml_goodput_measurement import goodput
 from ml_goodput_measurement import monitoring
+from praxis import trees
 
 # pylint: disable=too-many-positional-arguments
 
@@ -168,14 +170,14 @@ def write_metrics_to_tensorboard(writer, metrics, step, config, is_training=True
           f"completed step: {step}, seconds: {metrics['scalar']['perf/step_time_seconds']:.3f}, "
           f"TFLOP/s/device: {metrics['scalar']['perf/per_device_tflops_per_sec']:.3f}, "
           f"Tokens/s/device: {metrics['scalar']['perf/per_device_tokens_per_sec']:.3f}, "
-          f"total_weights: {metrics['scalar']['learning/total_weights']}, "
+          #f"total_weights: {metrics['scalar']['learning/total_weights']}, "
           f"loss: {metrics['scalar']['learning/loss']:.3f}"
       )
 
-      if full_log and jax.process_index() == 0:
-        max_logging.log(f"To see full metrics 'tensorboard --logdir={config.tensorboard_dir}'")
-        writer.flush()
 
+    if full_log and jax.process_index() == 0:
+      max_logging.log(f"To see full metrics 'tensorboard --logdir={config.tensorboard_dir}'")
+      writer.flush()
 
 def clear_buffered_metrics():
   global _buffered_step
@@ -299,9 +301,12 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
   xent = nn.with_logical_constraint(xent, ("activation_embed_and_logits_batch", "activation_length"))
   # Mask out paddings at the end of each example.
   xent = xent * (data["targets_segmentation"] != 0)
-  total_loss = jnp.sum(xent)
-  total_weights = jnp.sum(data["targets_segmentation"] != 0)
-  loss = total_loss / (total_weights + EPS)
+  per_batch_xent = jnp.sum(xent, axis=-1)
+  per_batch_weights = jnp.sum(data["targets_segmentation"] != 0, axis=-1)
+  per_batch_loss = per_batch_xent / (per_batch_weights + EPS)
+  total_loss = jnp.sum(per_batch_xent)
+  total_weights = jnp.sum(per_batch_weights)
+  loss = jnp.sum(per_batch_loss)
   # get moe load balance loss
   moe_lb_loss = 0.0
   if config.num_experts > 1:
@@ -318,6 +323,22 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
   return loss, aux
 
 
+
+# handling automatic microbatches via legate
+class MicrobatchConfig:
+  size: Optional[int] = None
+  schedule: str = "wavefront"
+  num_stages: Optional[int] = None
+  interleave: int = 1
+
+microbatch_config = MicrobatchConfig()
+
+def set_mb_config(mb_size, schedule, num_stages, interleave):
+    microbatch_config.size = mb_size
+    microbatch_config.schedule = schedule
+    microbatch_config.num_stages = num_stages
+    microbatch_config.interleave = interleave
+
 def train_step(model, config, state, data, dropout_rng):
   """
 
@@ -367,7 +388,28 @@ def train_step(model, config, state, data, dropout_rng):
     aux = jax.tree_map(lambda x: jnp.sum(x, axis=0), aux)
   else:
     grad_func = jax.value_and_grad(loss_fn, argnums=4, has_aux=True)
-    (loss, aux), raw_grads = grad_func(model, config, data, dropout_rng, state.params, is_train=True)
+
+    ## logging
+    if microbatch_config.size is None or microbatch_config.size == config.global_batch_size_to_train_on:
+      (loss, aux), raw_grads = grad_func(model, config, data, dropout_rng, state.params, is_train=True)
+    else:
+      from legate.jax import microbatch
+
+      def microbatch_input_sharding(x):
+        return None
+
+      arg_shardings = jax.tree_map(microbatch_input_sharding, data)
+
+      def mb_grad_func(data, dropout, params):
+        return grad_func(model, config, data, dropout, params, is_train=True)
+
+      mb_grad_func = microbatch(mb_grad_func, dim=0, argnum=0,
+                             size=microbatch_config.size, arg_shardings=arg_shardings,
+                             schedule=microbatch_config.schedule, num_stages=microbatch_config.num_stages,
+                             interleave=microbatch_config.interleave)
+
+      (loss, aux), raw_grads = mb_grad_func(data, dropout_rng, state.params)
+
   intermediate_outputs = aux["intermediate_outputs"]
   total_weights = aux["total_weights"]
   moe_lb_loss = aux["moe_lb_loss"]
@@ -380,11 +422,11 @@ def train_step(model, config, state, data, dropout_rng):
   metrics = {
       "scalar": {
           "learning/loss": loss,
-          "learning/moe_lb_loss": moe_lb_loss,
-          "learning/total_weights": total_weights,
-          "learning/grad_norm": max_utils.l2norm_pytree(grads),
-          "learning/raw_grad_norm": max_utils.l2norm_pytree(raw_grads),
-          "learning/param_norm": max_utils.l2norm_pytree(new_state.params),
+          #"learning/moe_lb_loss": moe_lb_loss,
+          #"learning/total_weights": total_weights,
+          #"learning/grad_norm": max_utils.l2norm_pytree(grads),
+          #"learning/raw_grad_norm": max_utils.l2norm_pytree(raw_grads),
+          #"learning/param_norm": max_utils.l2norm_pytree(new_state.params),
       },
       "scalars": {},
   }
@@ -530,15 +572,29 @@ def setup_train_loop(config):
   record_goodput(recorder, config, recorder.record_training_preparation_start_time if recorder else None)
   data_iterator, eval_data_iterator = create_data_iterator(config, mesh)
 
-  state, state_mesh_annotations, data_iterator = max_utils.setup_training_state(
-      model, data_iterator, tx, config, init_rng, mesh, checkpoint_manager
-  )
+  #state, state_mesh_annotations, data_iterator = max_utils.setup_training_state(
+  #    model, data_iterator, tx, config, init_rng, mesh, checkpoint_manager
+  #)
+
+  #if not config.using_pipeline_parallelism:
+  #  # The vocab tensor(s) of shape [vocab, embed] (and transpose) are not sharded by stage
+  #  maxtext_utils.assert_params_sufficiently_sharded(state.params, mesh, tolerance=0.02)
+  #record_goodput(recorder, config, recorder.record_training_preparation_end_time if recorder else None)
+
+  # Shaped RNG keys
+  _, example_rng = jax.random.split(jax.random.PRNGKey(0), 2)
+  shaped_rng = jax.ShapeDtypeStruct(example_rng.shape, example_rng.dtype)
+
+  # un-sharded abstract state
+  init_state_partial, abstract_state, state_mesh_annotations = max_utils.get_abstract_state_no_sharding(model, tx, config, example_rng, mesh)
+
+  # Shaped batch
+  shaped_batch = input_pipeline_interface.get_shaped_batch(config)
+
+  shaped_train_args = (abstract_state, shaped_batch, shaped_rng)
 
-  if not config.using_pipeline_parallelism:
-    # The vocab tensor(s) of shape [vocab, embed] (and transpose) are not sharded by stage
-    maxtext_utils.assert_params_sufficiently_sharded(state.params, mesh, tolerance=0.02)
-  record_goodput(recorder, config, recorder.record_training_preparation_end_time if recorder else None)
   return (
+      init_state_partial,
       init_rng,
       writer,
       checkpoint_manager,
@@ -548,7 +604,7 @@ def setup_train_loop(config):
       learning_rate_schedule,
       data_iterator,
       eval_data_iterator,
-      state,
+      shaped_train_args,
   )
 
 
@@ -565,6 +621,7 @@ def train_loop(config, state=None):
   record_goodput(recorder, config, recorder.record_job_start_time if recorder else None)
 
   (
+      init_state_partial,
       init_rng,
       writer,
       checkpoint_manager,
@@ -574,8 +631,9 @@ def train_loop(config, state=None):
       learning_rate_schedule,
       data_iterator,
       eval_data_iterator,
-      state,
+      shaped_train_args,
   ) = setup_train_loop(config)
+
   # pylint: disable=line-too-long
   (
       functional_train,
@@ -583,7 +641,7 @@ def train_loop(config, state=None):
       out_shard_train,
       static_argnums_train,
       donate_argnums_train,
-  ) = maxtext_utils.get_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config)
+  ) = maxtext_utils.get_and_compile_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config, shaped_train_args)
 
   if eval_data_iterator:
     # pylint: disable=line-too-long
@@ -595,6 +653,9 @@ def train_loop(config, state=None):
         donate_argnums_eval,
     ) = maxtext_utils.get_functional_eval_with_signature(eval_step, mesh, state_mesh_annotations, model, config)
 
+  # postponed initial state fetch as we needed to evaluate sharding first
+  state = max_utils.initialize_state_with_sharding(init_state_partial, mesh, in_shard_train[0], in_shard_train[2], init_rng)
+
   num_model_parameters = max_utils.calculate_num_params_from_pytree(state.params)
   max_logging.log(f"number parameters: {num_model_parameters/1e9:.3f} billion")
   per_device_tflops, _, _ = maxtext_utils.calculate_tflops_training_per_device(config)
@@ -614,15 +675,9 @@ def train_loop(config, state=None):
     p_eval_step = None
     print("Loaded compiled function!", flush=True)
   else:
-    p_train_step = jax.jit(
-        functional_train,
-        in_shardings=in_shard_train,
-        out_shardings=out_shard_train,
-        static_argnums=static_argnums_train,
-        donate_argnums=donate_argnums_train,
-    )
+    p_train_step=functional_train
 
-    if eval_data_iterator:
+    if config.eval_interval > 0 and eval_data_iterator:
       p_eval_step = jax.jit(
           functional_eval,
           in_shardings=in_shard_eval,
@@ -675,7 +730,8 @@ def train_loop(config, state=None):
         checkpoint_manager.wait_until_finished()
         sys.exit()
 
-    write_metrics(writer, local_metrics_file, running_gcs_metrics, metrics, step, config)
+    if writer:
+      write_metrics(writer, local_metrics_file, running_gcs_metrics, metrics, step, config)
 
     if config.eval_interval > 0 and step > start_step and (step + 1) % config.eval_interval == 0:
       assert eval_data_iterator
@@ -721,7 +777,8 @@ def train_loop(config, state=None):
 
   if checkpoint_manager is not None:
     checkpoint_manager.wait_until_finished()
-  write_metrics(writer, local_metrics_file, running_gcs_metrics, metrics, config.steps - 1, config)  # final step metrics
+  if writer:
+    write_metrics(writer, local_metrics_file, running_gcs_metrics, metrics, config.steps - 1, config)  # final step metrics
   max_utils.close_summary_writer(writer)
   record_goodput(recorder, config, recorder.record_job_end_time if recorder else None)
   clear_buffered_metrics()
@@ -736,6 +793,7 @@ def main(argv: Sequence[str]) -> None:
   pyconfig.initialize(argv)
   max_utils.print_system_information()
   config = pyconfig.config
+
   validate_train_config(config)
   os.environ["TFDS_DATA_DIR"] = config.dataset_path
   vertex_tensorboard_manager = VertexTensorboardManager()
