diff --git a/paxml/checkpoint_paths.py b/paxml/checkpoint_paths.py
index 5e0c343..bd73cc8 100644
--- a/paxml/checkpoint_paths.py
+++ b/paxml/checkpoint_paths.py
@@ -339,6 +339,20 @@ class PaxStepNameFormat(ocp.step.NameFormat):
     step = step or get_step_from_checkpoint_asset(step_path)
     return ocp.step.Metadata(step=step, path=step_path)

+  def find_step(
+      self, base_path: epath.PathLike, step: int
+  ) -> ocp.step.Metadata:
+    """Returns metadata for the given step."""
+    step_path = ocp.step.build_step_path(base_path, self, step)
+    metadata = self.build_metadata(step_path, step=step)
+    if metadata is not None:
+      return metadata
+
+    raise ValueError(
+        f'No step path found with name={self.build_name(step)},'
+        f' NameFormat={self} for step={step} under {base_path}.'
+    )
+
   def find_metadata(
       self, base_path: epath.PathLike, step: int
   ) -> ocp.step.Metadata | None:
diff --git a/paxml/contrib/gpu/scripts_gpu/configs.py b/paxml/contrib/gpu/scripts_gpu/configs.py
index 2aceb30..3be6de8 100644
--- a/paxml/contrib/gpu/scripts_gpu/configs.py
+++ b/paxml/contrib/gpu/scripts_gpu/configs.py
@@ -23,6 +23,7 @@ from paxml.contrib.gpu.scripts_gpu.llama_utils import BaseLLaMA
 from paxml.contrib.gpu.scripts_gpu.tasks import BoolQDataset
 from paxml.contrib.gpu.scripts_gpu.tasks import LambadaDataset
 from paxml.contrib.gpu.scripts_gpu.tasks import PileUnsupervisedDataset
+from paxml.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper
 from paxml.tasks.lm.params.c4 import TransformerLmSpmdAdam
 from paxml.tasks.lm.params.lm_cloud import SyntheticDataset
 from praxis import base_layer
@@ -109,9 +110,12 @@ class GPT126MBase(TransformerLmSpmdAdam):
   FPROP_DTYPE = jnp.bfloat16
   MAX_STEPS = 600000
 
+  USE_ADAFACTOR = False
+  USE_SGD = False
+
   MAX_SEQ_LEN = 2048
   VOCAB_SIZE = 50304
-  PACKED_INPUT = True
+  PACKED_INPUT = False
   PERCORE_BATCH_SIZE = 4
 
   NUM_LAYERS = 12
@@ -166,10 +170,12 @@ class GPT126MBase(TransformerLmSpmdAdam):
         fdl.get_callable(stacked_p), transformers.StackedTransformerRepeated
     ):
       stacked_p = stacked_p.block
-    transformer_layer_p = stacked_p.transformer_layer_params_tpl
-    transformer_layer_p.ln_tpl.reductions_in_fp32 = True
-    transformer_layer_p.tr_fflayer_tpl.ln_tpl.reductions_in_fp32 = True
+
     task_p.model.lm_tpl.final_ln_tpl.reductions_in_fp32 = True
+    if not TransformerEngineHelper.is_enabled_te():
+      transformer_layer_p = stacked_p.transformer_layer_params_tpl
+      transformer_layer_p.ln_tpl.reductions_in_fp32 = True
+      transformer_layer_p.tr_fflayer_tpl.ln_tpl.reductions_in_fp32 = True
 
     model_p.params_init = WeightInit.Gaussian(self.INIT_STD)
     softmax_init = WeightInit.Gaussian(self.SOFTMAX_INIT_STD)
@@ -180,6 +186,31 @@ class GPT126MBase(TransformerLmSpmdAdam):
     return task_p
 
 
+@experiment_registry.register
+class Pile126M(GPT126MBase, PileUnsupervisedDataset):
+
+  def task(self) -> pax_fiddle.Config[tasks_lib.SingleTask]:
+    task_p = super().task()
+    return task_p
+
+
+from paxml.tasks.lm.params.lm_cloud import SyntheticDataset
+@experiment_registry.register
+class Lambada126M(GPT126MBase, SyntheticDataset):
+
+  ICI_MESH_SHAPE = [8,1,1]
+  MAX_STEPS = 100
+  EVAL_INTERVAL_STEPS = 10
+
+  def task(self) -> pax_fiddle.Config[tasks_lib.SingleTask]:
+    task_p = super().task()
+    task_p.train.always_use_train_for_model_init=False
+    task_p.model.report_strict_acc=True
+    task_p.train.num_train_steps = self.MAX_STEPS
+    task_p.train.eval_interval_steps = self.EVAL_INTERVAL_STEPS
+    task_p.model.eval_task = 'lambada'
+    return task_p
+
 ## 32 node
 class GPT5BBase(GPT126MBase):
 
@@ -234,7 +265,7 @@ class GPT175BBase(GPT126MBase):
   # Known as MLP_DIM in t5x
   HIDDEN_DIMS = MODEL_DIMS * 4
   # Defaults to MODEL_DIMS // NUM_HEADS.
-  DIMS_PER_HEAD = None
+  DIMS_PER_HEAD = 128
   # Known as NUM_EMBEDDINGS in t5x
   VOCAB_SIZE = 50257
   USE_REPEATED_LAYER = True
@@ -306,15 +337,6 @@ class Synthetic175B(GPT175BBase, SyntheticDataset):
     return task_p
 
 
-### configs with the Pile dataset
-@experiment_registry.register
-class Pile126M(GPT126MBase, PileUnsupervisedDataset):
-
-  def task(self) -> pax_fiddle.Config[tasks_lib.SingleTask]:
-    task_p = super().task()
-    return task_p
-
-
 @experiment_registry.register
 class Pile5B(GPT5BBase, PileUnsupervisedDataset):
 
@@ -331,19 +353,6 @@ class Pile175B(GPT175BBase, PileUnsupervisedDataset):
     return task_p
 
 
-### example of a config that runs evaluation on the lambada dataset
-@experiment_registry.register
-class Lambada126M(GPT126MBase, LambadaDataset):
-
-  ICI_MESH_SHAPE = [8, 1, 1]
-
-  def task(self) -> pax_fiddle.Config[tasks_lib.SingleTask]:
-    task_p = super().task()
-    task_p.train.always_use_train_for_model_init = False
-    task_p.model.eval_task = 'lambada'
-    return task_p
-
-
 ### legacy aliases
 GPT5B = Pile5B
 GPT175B = Pile175B
diff --git a/paxml/contrib/gpu/scripts_gpu/tasks.py b/paxml/contrib/gpu/scripts_gpu/tasks.py
index 9c9ec60..bdc9b73 100644
--- a/paxml/contrib/gpu/scripts_gpu/tasks.py
+++ b/paxml/contrib/gpu/scripts_gpu/tasks.py
@@ -227,8 +227,8 @@ class BoolQDataset(base_experiment.BaseExperiment):
   EOS_ID: int = 2
 
   s = seqio.SentencePieceVocabulary(vocab_path)
-  TRUE_TOKEN: int = s.encode('yes')
-  FALSE_TOKEN: int = s.encode('no')
+  #TRUE_TOKEN: int = s.encode('yes')
+  #FALSE_TOKEN: int = s.encode('no')
 
   def _dataset_common(
       self, is_training
@@ -272,4 +272,4 @@ class BoolQDataset(base_experiment.BaseExperiment):
 
   def datasets(self) -> list[pax_fiddle.Config[base_input.BaseInput]]:
     """Returns a list of dataset parameters."""
-    return [self._dataset_common(is_training=False)]
+    return [self._dataset_common(is_training=self.LAMBADA_TRAIN)]
diff --git a/paxml/contrib/gpu/scripts_gpu/te_helper.py b/paxml/contrib/gpu/scripts_gpu/te_helper.py
new file mode 100644
index 0000000..cbac7cf
--- /dev/null
+++ b/paxml/contrib/gpu/scripts_gpu/te_helper.py
@@ -0,0 +1,76 @@
+import os
+from contextlib import contextmanager
+
+from praxis import base_layer
+
+try:
+  import transformer_engine.jax as te
+  from transformer_engine.common import recipe
+  _IS_TRANSFORMER_ENGINE_INSTALLED = True
+  DEFAULT_INIT_MUTABLE_LIST = base_layer.DEFAULT_INIT_MUTABLE_LIST + [te.fp8.FP8Helper.FP8_COLLECTION_NAME]
+
+except ModuleNotFoundError as e:
+  _IS_TRANSFORMER_ENGINE_INSTALLED = False
+  DEFAULT_INIT_MUTABLE_LIST = base_layer.DEFAULT_INIT_MUTABLE_LIST
+
+
+class TransformerEngineHelperBase:
+
+    @staticmethod
+    @contextmanager
+    def fp8_autocast(dp_mesh_axis="replica", tp_mesh_axis="mdl", fsdp_mesh_axis="data"):
+        raise NotImplementedError
+
+
+class TENotInstalledHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    @contextmanager
+    def fp8_autocast(dp_mesh_axis="replica", tp_mesh_axis="mdl", fsdp_mesh_axis="data"):
+        try:
+            yield
+        finally:
+            pass
+
+
+class TEInstalledHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    @contextmanager
+    def fp8_autocast(dp_mesh_axis="replica", tp_mesh_axis="mdl", fsdp_mesh_axis="data"):
+        fp8_recipe = recipe.DelayedScaling(margin=0, interval=1, fp8_format=recipe.Format.HYBRID,
+                                           amax_history_len=1024, amax_compute_algo='max')
+
+        enable_fp8 = bool(int((os.environ.get("ENABLE_FP8", False))))
+        try:
+            with te.fp8_autocast(enabled=enable_fp8,
+                                 fp8_recipe=fp8_recipe,
+                                 mesh_resource=te.MeshResource(dp_resource=dp_mesh_axis,
+                                                               tp_resource=tp_mesh_axis,
+                                                               fsdp_resource=fsdp_mesh_axis)):
+                yield
+        finally:
+            pass
+
+
+class TransformerEngineHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def is_enabled_te():
+        enable_te = bool(int((os.environ.get("ENABLE_TE", False))))
+        return (_IS_TRANSFORMER_ENGINE_INSTALLED and enable_te)
+
+    @staticmethod
+    def get_helper():
+        if TransformerEngineHelper.is_enabled_te():
+            return TEInstalledHelper
+        return TENotInstalledHelper
+
+    @staticmethod
+    @contextmanager
+    def fp8_autocast(dp_mesh_axis="replica", tp_mesh_axis="mdl", fsdp_mesh_axis="data"):
+        try:
+            with TransformerEngineHelper.get_helper().fp8_autocast(dp_mesh_axis, tp_mesh_axis, fsdp_mesh_axis):
+                yield
+        finally:
+            pass
diff --git a/paxml/executors.py b/paxml/executors.py
index 26cacbb..43d67e0 100644
--- a/paxml/executors.py
+++ b/paxml/executors.py
@@ -395,13 +395,13 @@ def _train_and_evaluate_common(
   gc.freeze()
   while True:
     logging.log_first_n(INFO, '[PAX STATUS]: Beginning step `%d`.', 5, step_i)
-    checkpointer.save_if_needed(
-        step_i,
-        partitioned_train_state,
-        train_state_metadata.unpadded_global_shapes,
-        train_state_metadata.partition_specs,
-        train_input_for_checkpoint,
-    )
+    #checkpointer.save_if_needed(
+    #    step_i,
+    #    partitioned_train_state,
+    #    train_state_metadata.unpadded_global_shapes,
+    #    train_state_metadata.partition_specs,
+    #    train_input_for_checkpoint,
+    #)
     if exit_after_ondemand_checkpoint and checkpointer.reached_preemption(
         step_i
     ):
@@ -542,13 +542,14 @@ def _train_and_evaluate_common(
   gc.unfreeze()
 
   logging.info('[PAX STATUS]: Saving checkpoint for final step.')
-  checkpointer.save_final(
-      step_i,
-      partitioned_train_state=partitioned_train_state,
-      train_state_unpadded_shape_dtype_struct=train_state_metadata.unpadded_global_shapes,
-      train_state_pspecs=train_state_metadata.partition_specs,
-      train_input_pipeline=train_input_for_checkpoint,
-  )
+  #checkpointer.save_final(
+  #    step_i,
+  #    partitioned_train_state=partitioned_train_state,
+  #    train_state_unpadded_shape_dtype_struct=train_state_metadata.unpadded_global_shapes,
+  #    train_state_pspecs=train_state_metadata.partition_specs,
+  #    train_input_pipeline=train_input_for_checkpoint,
+  #)
+  #
+  #checkpointer.close()
 
-  checkpointer.close()
   logging.info('[PAX STATUS]: Final checkpoint saved.')
diff --git a/paxml/learners.py b/paxml/learners.py
index a5eb356..563275c 100644
--- a/paxml/learners.py
+++ b/paxml/learners.py
@@ -130,9 +130,9 @@ class Learner(base_hyperparams.FiddleBaseParameterizable):
   loss_name: str | None = None
   stochastic_gradient: sgf.BaseStochasticGradient | None = None
   skip_zero_gradients: bool | None = None
-  grad_norm_summary: bool = True
+  grad_norm_summary: bool = False
   grad_norm_individual_vars: bool = False
-  var_norm_summary: bool = True
+  var_norm_summary: bool = False
   var_norm_individual_vars: bool = False
   check_valid_step: bool = True
   vectorize_on_repeat_prefix: bool = True
diff --git a/paxml/main.py b/paxml/main.py
index ff3d89d..035d1b3 100644
--- a/paxml/main.py
+++ b/paxml/main.py
@@ -22,6 +22,11 @@ python paxml/main.py \
 """
 # Internal import for aiding module import speed
 
+import tensorflow as tf
+tf.config.threading.set_intra_op_parallelism_threads(1)
+tf.config.threading.set_inter_op_parallelism_threads(1)
+tf.config.set_soft_device_placement(True)
+
 import contextlib
 import importlib
 import os
@@ -51,6 +56,7 @@ from paxml import tf_data_service_lib
 from paxml import train
 from paxml import trainer_lib
 from paxml import tuning_lib
+from paxml.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper
 from praxis import pax_fiddle
 from praxis import py_utils
 
@@ -58,7 +64,6 @@ from praxis import py_utils
 # internal experiment module import
 # internal import for g3_multiprocessing
 
-
 FLAGS = flags.FLAGS
 
 flags.DEFINE_string(
@@ -512,39 +517,41 @@ def _main(argv: Sequence[str]) -> None:
                                                       FLAGS.host_idx)
                      )
 
-  if FLAGS.exp is not None:
-    experiment_config = get_experiment(FLAGS.exp)()
-  elif absl_flags.fdl_flags_supplied():
-    # Use the legacy Fiddle flags API to parse command line Fiddle flags.
-    cfg = absl_flags.create_buildable_from_flags(
-        module=None, allow_imports=True)
-    experiment_config = pax_fiddle.build(cfg)
-    logging.warning(
-        'Legacy Fiddle flags API usage detected. Please use the new Fiddle'
-        ' command line flag `fdl` with various commands to specify the'
-        ' config and any overrides. Please see'
-        ' `fiddle/docs/flags_code_lab.md` for more'
-        ' documentation on Fiddle flags usage.'
-    )
-  elif _FIDDLE_CONFIG.value is not None:
-    # This uses the new Fiddle flags API `DEFINE_fiddle_config()` to parse
-    # command line Fiddle flags. See
-    # `fiddle/docs/flags_code_lab.md` for details on the new
-    # Fiddle flags API.
-    logging.info(
-        'Using pax_fiddle_config from the command line: %s',
-        _FIDDLE_CONFIG.value,
-    )
-    experiment_config = pax_fiddle.build(_FIDDLE_CONFIG.value)
-  else:
-    raise app.UsageError(
-        'No experiment provided. At least one of --exp, --fdl,'
-        ' --fdl_config, or --fdl_config_file is required.'
-    )
+  with TransformerEngineHelper.fp8_autocast('replica', 'mdl', 'data'):
+    if FLAGS.exp is not None:
+      experiment_config = get_experiment(FLAGS.exp)()
+    elif absl_flags.fdl_flags_supplied():
+      # Use the legacy Fiddle flags API to parse command line Fiddle flags.
+      cfg = absl_flags.create_buildable_from_flags(
+          module=None, allow_imports=True)
+      experiment_config = pax_fiddle.build(cfg)
+      logging.warning(
+          'Legacy Fiddle flags API usage detected. Please use the new Fiddle'
+          ' command line flag `fdl` with various commands to specify the'
+          ' config and any overrides. Please see'
+          ' `fiddle/docs/flags_code_lab.md` for more'
+          ' documentation on Fiddle flags usage.'
+      )
+    elif _FIDDLE_CONFIG.value is not None:
+      # This uses the new Fiddle flags API `DEFINE_fiddle_config()` to parse
+      # command line Fiddle flags. See
+      # `fiddle/docs/flags_code_lab.md` for details on the new
+      # Fiddle flags API.
+      logging.info(
+          'Using pax_fiddle_config from the command line: %s',
+          _FIDDLE_CONFIG.value,
+      )
+      experiment_config = pax_fiddle.build(_FIDDLE_CONFIG.value)
+    else:
+      raise app.UsageError(
+          'No experiment provided. At least one of --exp, --fdl,'
+          ' --fdl_config, or --fdl_config_file is required.'
+      )
+
 
-  experiment_config.validate()
-  run(experiment_config=experiment_config,
-      enable_checkpoint_saving=FLAGS.enable_checkpoint_saving)
+    experiment_config.validate()
+    run(experiment_config=experiment_config,
+        enable_checkpoint_saving=FLAGS.enable_checkpoint_saving)
 
 
 _TASK_HANDLE_RE = re.compile(r'(?:logs\.)?(\d+)\.(.*)\.([^.]+)\.\d+')
diff --git a/paxml/partitioning.py b/paxml/partitioning.py
index 7deb362..69c4a54 100644
--- a/paxml/partitioning.py
+++ b/paxml/partitioning.py
@@ -20,7 +20,7 @@ import dataclasses
 import functools
 import json
 import pprint
-from typing import Any, Protocol, Sequence
+from typing import Any, Optional, Protocol, Sequence
 
 from absl import logging
 from clu import platform
@@ -42,6 +42,7 @@ from praxis import pax_fiddle
 from praxis import py_utils
 from praxis import pytypes
 from praxis import trees
+import legate.jax
 
 # tf_data_service_lib is slow to import, so we do it lazily.
 tf_data_service_lib = lazy_loader.LazyLoader(
@@ -64,6 +65,112 @@ TrainStateProvenance = train_states.TrainStateProvenance
 TrainStateMetadata = trainer_lib.TrainStateMetadata
 RunningMode = trainer_lib.RunningMode
 
+import contextlib
+from jax._src.mesh import thread_resources
+from jax._src import config as jax_config
+import enum
+from collections import OrderedDict
+import numpy as np
+
+class LegateMeshWrapper(contextlib.ContextDecorator):
+
+  class Mode(enum.IntEnum):
+    LOWERING = 0
+    COMPILING = 1
+    NONE = 2
+  _mode = Mode.NONE
+
+  def __init__(self, global_mesh, local_mesh: tuple[int]):
+    self.global_mesh = global_mesh
+    self.local_shape = OrderedDict( (name, dim) for name, dim in self.global_mesh.shape.items() )
+    for local_dim, (name, global_dim) in zip(local_mesh, self.global_mesh.shape.items()):
+      if local_dim != global_dim:
+        self.local_shape[name] = local_dim
+
+    self.num_local_devices = np.prod(local_mesh)
+
+    local_devices_slice = tuple( slice(0,dim) for dim in local_mesh )
+    self._local_devices = self.global_mesh.devices[local_devices_slice]
+
+
+  @classmethod
+  @contextlib.contextmanager
+  def mode(cls, mode: Mode):
+    save = cls._mode
+    cls._mode = mode
+    yield
+    cls._mode = save
+
+  def __repr__(self):
+    return f"LegateMeshWrapper({repr(self.global_mesh)})"
+
+  def __str__(self):
+    return f"LegateMeshWrapper({str(self.global_mesh)})"
+
+  @property
+  def is_multi_process(self):
+    return self.global_mesh.is_multi_process
+
+  @property
+  def _flat_devices_tuple(self):
+    devices = self.global_mesh._flat_devices_tuple
+    if self.compiling():
+      return devices[:self.num_local_devices]
+    return devices
+
+  @property
+  def local_devices(self):
+    return self.global_mesh.local_devices
+
+  def compiling(self) -> bool:
+    return LegateMeshWrapper._mode == self.Mode.COMPILING
+
+  @property
+  def axis_names(self):
+    return self.global_mesh.axis_names
+
+  @property
+  def devices(self):
+    if self.compiling():
+      return self._local_devices
+    return self.global_mesh.devices
+
+  @property
+  def shape(self):
+    if self.compiling():
+      return self.local_shape
+    return self.global_mesh.shape
+
+  @property
+  def _internal_device_list(self):
+    return self.global_mesh._internal_device_list
+
+  @property
+  def size(self):
+    if self.compiling():
+      return self.num_local_devices
+    return self.global_mesh.size
+
+  @property
+  def empty(self):
+    return self.global_mesh.empty
+
+  def __enter__(self):
+    new_env = thread_resources.stack[-1].with_mesh(self)
+    thread_resources.stack.append(new_env)
+    thread_resources.env = new_env
+    jax_config.update_thread_local_jit_state(
+        mesh_context_manager=tuple(t.physical_mesh for t in thread_resources.stack
+                                  if not t.physical_mesh.empty))
+    return self
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    thread_resources.stack.pop()
+    thread_resources.env = thread_resources.stack[-1]
+    jax_config.update_thread_local_jit_state(
+        mesh_context_manager=tuple(t.physical_mesh for t in thread_resources.stack
+                                  if not t.physical_mesh.empty))
+    return False
 
 def _identity(x):
   """A helper identity function, defined globally so it is JIT-compiled once."""
@@ -127,9 +234,24 @@ def compile_for_auto_sharding(
     return core.ShapedArray(x.shape, dtype)
 
   inputs_shape_dtype = jax.tree.map(_create_aval, inputs_shape_dtype)
-  compiled = step_fn.lower(
-      train_state, step_key, inputs_shape_dtype, static_args
-  ).compile()
+  with LegateMeshWrapper.mode(LegateMeshWrapper.Mode.LOWERING):
+    lowered = step_fn.lower(
+        train_state, step_key, inputs_shape_dtype, static_args
+    )
+  with LegateMeshWrapper.mode(LegateMeshWrapper.Mode.COMPILING):
+    compiled = lowered.compile()
+
+  sharding_str_arr = []
+  def check(inp, inp_sharding, out):
+    sharding_str_arr.append(f"{inp.shape} {inp.sharding} {out}")
+    if not inp_sharding.is_equivalent_to(out, len(inp.shape)):
+      raise Exception("\n".join(sharding_str_arr) + " mismatch")
+
+  # validate that the donated input parameters and their aliased outputs
+  # have the same shardings so that after updating the input the sharding
+  # is the same across iterations
+  jax.tree_map(check, train_state, compiled.input_shardings[0][0], compiled.output_shardings[0])
+
   return compiled, compiled.input_shardings[0]
 
 
@@ -740,6 +862,7 @@ class PmapPartitioner(Partitioner):
     ):
       if static_args:
         static_args = static_args.replace(unpadded_global_batch_size=None)
+      config = trainer_lib.PaxLegateConfig()
       return partitioned_step_fn(state, prng_key, inputs, static_args)
 
     return _wrapped_partitioned_step, None  # Input partition spec.
@@ -782,7 +905,10 @@ class PjitPartitioner(Partitioner):
     else:
       logging.info('Using provided mesh for PjitPartitioner')
     logging.info('device_mesh: %s', device_mesh)
-    self._global_mesh = jax.sharding.Mesh(device_mesh, model.mesh_axis_names)
+
+    config = trainer_lib.PaxLegateConfig()
+
+    self._global_mesh = LegateMeshWrapper(jax.sharding.Mesh(device_mesh, model.mesh_axis_names), config.local_mesh)
 
     # Pjit'ed function to preprocess the prng key.
     self._broadcast_key_fn = None
@@ -894,9 +1020,17 @@ class PjitPartitioner(Partitioner):
   ) -> NestedJTensor:
     """Preprocess the input batch before using it."""
     if self._reshard_inputs:
-      return input_pipeline.reshard_for_spmd(
-          padded_inputs, self.global_mesh, partition_specs
+      global_shapes = jax.tree_util.tree_map(
+          py_utils.get_global_input_shape_dtype, padded_inputs
       )
+
+      def _make_array(x, global_shape, sharding):
+        if isinstance(sharding, PartitionSpec):
+          sharding = jax.sharding.NamedSharding(self.global_mesh, sharding)
+        return py_utils.make_array(x, global_shape, self.global_mesh, pspecs=None, sharding=sharding)
+
+      return jax.tree_util.tree_map(_make_array, padded_inputs, global_shapes, partition_specs)
+
     return padded_inputs
 
   def get_train_state_metadata(
@@ -1009,7 +1143,8 @@ class PjitPartitioner(Partitioner):
       # need to keep this resharding, because the resharding is related to the
       # locations of the real data (inputs_split_mapping) and is not visible
       # from HLOs.
-      inputs = jax.tree.map(reshard_inputs_fn, inputs)
+      #inputs = jax.tree_map(reshard_inputs_fn, inputs)
+
 
       fn_out = step_fn(
           self._jax_task,
@@ -1043,18 +1178,14 @@ class PjitPartitioner(Partitioner):
     logging.info('step_fn fn_in_partition_specs=%s', fn_in_partition_specs)
     logging.info('step_fn fn_out_partition_specs=%s', fn_out_partition_specs)
 
-    fn_in_shardings = jax.tree.map(
-        lambda p: jax.sharding.NamedSharding(self._global_mesh, p)
-        if not isinstance(p, pjit.AUTO)
-        else p,
-        fn_in_partition_specs,
-    )
-    fn_out_shardings = jax.tree.map(
-        lambda p: jax.sharding.NamedSharding(self._global_mesh, p)
-        if not isinstance(p, pjit.AUTO)
-        else p,
-        fn_out_partition_specs,
-    )
+    def canonicalize_sharding(p):
+      if isinstance(p, pjit.AUTO) or isinstance(p, jax.sharding.NamedSharding) or isinstance(p, jax.sharding.GSPMDSharding):
+        return p
+      return jax.sharding.NamedSharding(self._global_mesh, p)
+
+    fn_in_shardings = jax.tree_map(canonicalize_sharding, fn_in_partition_specs)
+    fn_out_shardings = jax.tree_map(canonicalize_sharding, fn_out_partition_specs)
+
     extra_kwargs = dict(in_shardings=fn_in_shardings)
     if not use_pspec_on_array_inputs:
       extra_kwargs = {}
@@ -1172,7 +1303,15 @@ class PjitPartitioner(Partitioner):
         fn_out_partition_specs,
         use_pspec_on_array_inputs=use_pspec_on_array_inputs,
     )
-    return partitioned_step_fn
+    def _wrapped_partitioned_step(
+      state,
+      prng_key,
+      inputs,
+      static_args: BaseStepFnStaticArgs | None = None,
+    ):
+      return partitioned_step_fn(state, prng_key, inputs, static_args)
+
+    return _wrapped_partitioned_step
 
 
 class AutoShardingPjitPartitioner(PjitPartitioner):
@@ -1296,8 +1435,13 @@ class AutoShardingPjitPartitioner(PjitPartitioner):
         else pjit.AUTO(self.global_mesh)
     )
 
+    from jax.lax import with_sharding_constraint
+    def autoshard_step_fn(train_state, *args):
+      sharded_train_state = jax.tree_map(lambda x, y: with_sharding_constraint(x, y), train_state, metadata.partition_specs)
+      return step_fn(sharded_train_state, *args)
+
     partitioned_step_fn = self._pjit(
-        step_fn,
+        autoshard_step_fn,
         is_eval,
         fn_in_partition_specs,
         fn_out_partition_specs,
@@ -1314,18 +1458,24 @@ class AutoShardingPjitPartitioner(PjitPartitioner):
         jax.tree.map(jnp.shape, inputs_shape_dtype),
         static_args,
     )
-    (
-        auto_sharded_step_fn,
-        input_shardings,
-    ) = compile_for_auto_sharding(
-        partitioned_step_fn,
-        metadata.unpadded_global_shapes,
-        self._init_key,
-        inputs_shape_dtype,
-        static_args,
-    )
-    new_train_state_pspec = jax.tree.map(lambda x: x.spec, input_shardings[0])
-    new_input_pspec = jax.tree.map(lambda x: x.spec, input_shardings[2])
+
+    config = trainer_lib.PaxLegateConfig()
+    with legate.jax.context(configurable=config.configurable, autoshard=True):
+      (
+          auto_sharded_step_fn,
+          input_shardings,
+      ) = compile_for_auto_sharding(
+          partitioned_step_fn,
+          metadata.unpadded_global_shapes,
+          self._init_key,
+          inputs_shape_dtype,
+          static_args,
+      )
+
+
+
+    new_train_state_pspec = input_shardings[0]
+    new_input_pspec = input_shardings[2]
     return auto_sharded_step_fn, new_input_pspec, new_train_state_pspec
 
   def get_train_state_metadata(
@@ -1408,6 +1558,7 @@ class AutoShardingPjitPartitioner(PjitPartitioner):
               subset=self._auto_sharding_input_spec, superset=inputs
           )
       )
+
       return partitioned_step_fn(state, prng_key, subset_extracted_inputs)
 
     self._auto_sharding_result = (
@@ -1440,6 +1591,12 @@ class AutoShardingPjitPartitioner(PjitPartitioner):
     if not self._auto_sharding_result:
       self.get_train_state_metadata(discard_opt_states=is_eval)
     if step_fn is self._auto_sharding_info.step_fn:
+      return (
+          self._auto_sharding_result.partitioned_step_fn,
+          self._auto_sharding_result.input_partition_spec,
+      )
+      # this sanity check appears to be wrong and compares global sizes
+      # against per-node sizes
       if trees.is_subset(
           self._auto_sharding_result.inputs_shape_dtype, inputs_shape_dtype
       ):
diff --git a/paxml/summary_utils.py b/paxml/summary_utils.py
index 0cc6bdd..4959aa5 100644
--- a/paxml/summary_utils.py
+++ b/paxml/summary_utils.py
@@ -821,6 +821,7 @@ class SummaryHandler:
     Returns:
       True if the summaries were written, False otherwise.
     """
+    return False
     should_accumulate = self.should_accumulate(step)
     should_write_summary = self.should_write(step)
     should_log = self.should_log(step)
diff --git a/paxml/tasks/lm/model_params.py b/paxml/tasks/lm/model_params.py
index ea54531..1179e76 100644
--- a/paxml/tasks/lm/model_params.py
+++ b/paxml/tasks/lm/model_params.py
@@ -105,13 +105,16 @@ def set_sharding_annotations_v1(
     dcn_mesh_shape: a 3D sequence representing the mesh across slices, or None.
   """
   model_p = task_p.model
+  # asserts.eq(len(ici_mesh_shape), 3)
   model_p.ici_mesh_shape = ici_mesh_shape
   if dcn_mesh_shape is not None:
+    # asserts.eq(len(dcn_mesh_shape), 3)
     model_p.dcn_mesh_shape = dcn_mesh_shape
   replica_axis = 'replica'
   data_axis = 'data'
   mdl_axis = 'mdl'
-  mesh_axis_names = [replica_axis, data_axis, mdl_axis]
+  seq_axis = 'seq'
+  mesh_axis_names = [replica_axis, data_axis, mdl_axis, seq_axis]
   task_p.train.inputs_split_mapping = NestedMap(
       map_1d=((replica_axis, data_axis),),
       map_2d=((replica_axis, data_axis), None))
@@ -125,6 +128,7 @@ def set_sharding_annotations_v1(
         replica_axis=replica_axis,
         data_axis=data_axis,
         mdl_axis=mdl_axis,
+        seq_axis=seq_axis,
         ici_mesh_shape=model_p.ici_mesh_shape,
         dcn_mesh_shape=model_p.dcn_mesh_shape,
         mesh_axis_names=mesh_axis_names,
@@ -583,6 +587,7 @@ class TransformerLmSpmdAdafactor(base_experiment.BaseExperiment):
   DECAY_END = 100000
   USE_FP8 = False
   USE_EXPERT_PARALLEL = False
+  REMAT = False
 
   # optimizer related
   DROPOUT_PROB = 0.0
@@ -671,6 +676,9 @@ class TransformerLmSpmdAdafactor(base_experiment.BaseExperiment):
     transformer_layer_p.tr_atten_tpl.use_bias = False
     transformer_layer_p.tr_atten_tpl.combine_qkv = self.COMBINE_QKV
 
+    stacked_transformer_tpl.checkpoint_policy = self.CHECKPOINT_POLICY
+    stacked_transformer_tpl.remat = self.REMAT
+
     if self.USE_FP8:
       transformer_layer_p.tr_atten_tpl.proj_tpl.einsum_tpl = pax_fiddle.Config(
           fp8_ops.Fp8EinsumOp
@@ -710,6 +718,7 @@ class TransformerLmSpmdAdafactor(base_experiment.BaseExperiment):
           self.CHECKPOINT_POLICY)
     else:
       model_p.lm_tpl.stacked_transformer_tpl = stacked_transformer_tpl
+      model_p.lm_tpl.stacked_transformer_tpl.checkpoint_policy = self.CHECKPOINT_POLICY
 
     # Enable bf16.
     model_p.fprop_dtype = self.FPROP_DTYPE
@@ -908,6 +917,7 @@ class TransformerLmSpmdPipelineAdafactor(TransformerLmSpmdAdafactor):
       stacked_transformer_tpl.x_times = self.NUM_LAYERS // (
           self.NUM_STAGES * self.CIRCULAR_REPEAT)
       stacked_transformer_tpl.checkpoint_policy = self.CHECKPOINT_POLICY
+      stacked_transformer_tpl.remat = self.REMAT
 
     # Wrap it with a pipeline layer.
     model_p.lm_tpl.stacked_transformer_tpl = pax_fiddle.Config(
diff --git a/paxml/tasks/lm/params/c4.py b/paxml/tasks/lm/params/c4.py
index 09d6f1e..35e525a 100644
--- a/paxml/tasks/lm/params/c4.py
+++ b/paxml/tasks/lm/params/c4.py
@@ -244,20 +244,35 @@ def set_adam_and_learning_rate_schedule(
   """Sets the Adam optimizer and the learning rate schedule."""
   lp = task_p.train.learner
   lp.loss_name = 'total_loss'
-  lp.optimizer = pax_fiddle.Config(
-      optimizers.Adam,
-      beta1=cls.ADAM_BETA1 if cls.ADAM_BETA1 else 0.9,
-      beta2=cls.ADAM_BETA2 if cls.ADAM_BETA2 else 0.999,
-      weight_decay=cls.WEIGHT_DECAY if cls.WEIGHT_DECAY else 0.0,
-      epsilon=cls.ADAM_EPSILON if cls.ADAM_EPSILON else 1e-6,
-      epsilon_root=cls.ADAM_EPSILON_ROOT if cls.ADAM_EPSILON_ROOT else 0.0,
-      clip_gradient_norm_to_value=cls.CLIP_GRADIENT_NORM_TO_VALUE
-      if cls.CLIP_GRADIENT_NORM_TO_VALUE
-      else 5.0,
-      clip_threshold=cls.ADAM_CLIP_THRESHOLD
-      if cls.ADAM_CLIP_THRESHOLD
-      else 1.0,
-  )
+
+  if cls.USE_ADAFACTOR:
+    lp.optimizer = pax_fiddle.Config(
+        optimizers.ShardedAdafactor,
+        decay_method='adam',
+        beta1=cls.ADAM_BETA1,
+        decay_adam=0.99,
+        weight_decay=cls.WEIGHT_DECAY,
+        clip_gradient_norm_to_value=cls.CLIP_GRADIENT_NORM_TO_VALUE,
+    )
+  elif cls.USE_SGD:
+    lp.optimizer = pax_fiddle.Config(
+      optimizers.Sgd,
+    )
+  else:
+    lp.optimizer = pax_fiddle.Config(
+        optimizers.Adam,
+        beta1=cls.ADAM_BETA1 if cls.ADAM_BETA1 else 0.9,
+        beta2=cls.ADAM_BETA2 if cls.ADAM_BETA2 else 0.999,
+        weight_decay=cls.WEIGHT_DECAY if cls.WEIGHT_DECAY else 0.0,
+        epsilon=cls.ADAM_EPSILON if cls.ADAM_EPSILON else 1e-6,
+        epsilon_root=cls.ADAM_EPSILON_ROOT if cls.ADAM_EPSILON_ROOT else 0.0,
+        clip_gradient_norm_to_value=cls.CLIP_GRADIENT_NORM_TO_VALUE
+        if cls.CLIP_GRADIENT_NORM_TO_VALUE
+        else 5.0,
+        clip_threshold=cls.ADAM_CLIP_THRESHOLD
+        if cls.ADAM_CLIP_THRESHOLD
+        else 1.0,
+    )
 
   if hasattr(cls, 'PERCORE_BATCH_SIZE'):
     global_batch_size = int(cls.PERCORE_BATCH_SIZE * jax.device_count() + 1e-6)
diff --git a/paxml/tasks/lm/params/nvidia.py b/paxml/tasks/lm/params/nvidia.py
index 24fa3b7..233ab63 100644
--- a/paxml/tasks/lm/params/nvidia.py
+++ b/paxml/tasks/lm/params/nvidia.py
@@ -33,7 +33,7 @@ from praxis import schedules
 from praxis.layers import activations
 from praxis.layers import glam
 from praxis.layers import gpu_fast_attention
-from praxis.layers import grok
+#from praxis.layers import grok
 from praxis.layers import transformers
 
 WeightInit = base_layer.WeightInit
@@ -46,7 +46,7 @@ class NVIDIA1_3B(c4.TransformerLmSpmdAdam, lm_cloud.SyntheticDataset):
   USE_FLASH_ATTENTION = False
   USE_TRITON_LAYER_NORM = False
 
-  USE_REPEATED_LAYER = True
+  USE_REPEATED_LAYER = False
 
   DCN_MESH_SHAPE = [1, 1, 1]
   ICI_MESH_SHAPE = [16, 1, 1]
@@ -64,6 +64,7 @@ class NVIDIA1_3B(c4.TransformerLmSpmdAdam, lm_cloud.SyntheticDataset):
   ACTIVATION_CLS = layers.GELU
   USE_GATED_ACTIVATION = False
   PACKED_INPUT = True
+  MAX_STEPS = 10
 
   FPROP_DTYPE = jnp.bfloat16
 
@@ -82,6 +83,8 @@ class NVIDIA1_3B(c4.TransformerLmSpmdAdam, lm_cloud.SyntheticDataset):
   CLIP_GRADIENT_NORM_TO_VALUE = 1.0
   CLIP_THRESHOLD = 1.0
 
+  USE_SGD = False
+
   # Learning rate schedule
   LR_SCHEDULE = 'linear_rampup_cosine_decay'
   LR_COS_WARMUP = 0
@@ -99,6 +102,8 @@ class NVIDIA1_3B(c4.TransformerLmSpmdAdam, lm_cloud.SyntheticDataset):
     # (Practically) disable summary writes.
     task_p.train.save_interval_steps = 100000
 
+    task_p.train.num_train_steps = self.MAX_STEPS
+
     model_p = task_p.model
     model_p.params_init = WeightInit.Gaussian(self.INIT_STD)
 
@@ -123,6 +128,12 @@ class NVIDIA1_3B(c4.TransformerLmSpmdAdam, lm_cloud.SyntheticDataset):
       fused_ln_tpl.copy_fields_from(layer_p.ln_tpl)
       layer_p.ln_tpl = fused_ln_tpl
 
+    model_p.lm_tpl.stacked_transformer_tpl.transformer_layer_params_tpl.tr_atten_tpl.atten_logit_cap = 0.
+    model_p.lm_tpl.stacked_transformer_tpl.transformer_layer_params_tpl.tr_atten_tpl.internal_enable_per_dim_scale = False
+    model_p.lm_tpl.stacked_transformer_tpl.transformer_layer_params_tpl.tr_fflayer_tpl.has_bias = False
+    model_p.lm_tpl.stacked_transformer_tpl.transformer_layer_params_tpl
+    model_p.lm_tpl.packed_input = False
+
     scale = self.SOFTMAX_INIT_STD
     if not scale:
       scale = 1.0 / math.sqrt(self.MODEL_DIMS)
diff --git a/paxml/tasks_lib.py b/paxml/tasks_lib.py
index cfdc402..e475099 100644
--- a/paxml/tasks_lib.py
+++ b/paxml/tasks_lib.py
@@ -43,6 +43,7 @@ from paxml import checkpoint_types
 from paxml import io_utils
 from paxml import learners as learners_lib
 from paxml import train_states
+from paxml.contrib.gpu.scripts_gpu.te_helper import DEFAULT_INIT_MUTABLE_LIST
 from praxis import asserts
 from praxis import base_hyperparams
 from praxis import base_input
@@ -1786,7 +1787,7 @@ class SingleTask(base_task.BaseTask):
       )
     # Initialize with a dummy seed
     var_weight_hparams = ckpt_task.model.abstract_init_with_metadata(
-        inputs_shape_dtype)
+        inputs_shape_dtype, mutable=DEFAULT_INIT_MUTABLE_LIST)
     ckpt_train_state = ckpt_task.create_train_state_padded_shapes(
         var_weight_hparams)
     train_state_pspecs = ckpt_task.create_train_state_partition_specs(
diff --git a/paxml/trainer_lib.py b/paxml/trainer_lib.py
index 6ec25e8..87fc8e0 100644
--- a/paxml/trainer_lib.py
+++ b/paxml/trainer_lib.py
@@ -19,7 +19,7 @@ import dataclasses
 import enum
 import functools
 import pprint
-from typing import Any, Protocol, Sequence
+from typing import Any, Protocol, Optional, Sequence
 
 from absl import logging
 import clu.metrics
@@ -29,12 +29,14 @@ from flax import struct as flax_struct
 import jax
 from jax import numpy as jnp
 from jax.experimental import pjit
+from jax.sharding import PartitionSpec as P
 from paxml import base_metrics
 from paxml import checkpoint_types
 from paxml import learners as learners_lib
 from paxml import sgf
 from paxml import tasks_lib
 from paxml import train_states
+from paxml.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper, DEFAULT_INIT_MUTABLE_LIST
 from praxis import asserts
 from praxis import base_hyperparams
 from praxis import base_input
@@ -90,6 +92,14 @@ PMAP_PARALLEL_AXIS_NAME = base_layer.PMAP_PARALLEL_AXIS_NAME
 instantiate = base_hyperparams.instantiate
 
 
+import gin
+import legate.jax
+
+@gin.configurable
+@dataclasses.dataclass
+class PaxLegateConfig(legate.jax.ClientConfig):
+  local_mesh: Optional[tuple[int]] = None
+
 class RunningMode(enum.Flag):
   """Running mode."""
 
@@ -167,8 +177,7 @@ def create_train_state_metadata(
     A TrainStateMetadata instance.
   """
   var_weight_hparams = jax_task.model.abstract_init_with_metadata(
-      train_shape_dtype, do_eval=do_eval
-  )
+      train_shape_dtype, do_eval=do_eval, extra_mutable_list=DEFAULT_INIT_MUTABLE_LIST)
   padded_global_shapes = jax_task.create_train_state_padded_shapes(
       var_weight_hparams, discard_opt_states=discard_opt_states
   )
@@ -217,7 +226,8 @@ def write_post_init_model_hparams_file(
     logging.info('post_init_model_params: %s', params_fpath)
     job_log_dir.mkdir(parents=True, exist_ok=True)
     hyper_params = model.abstract_init_with_mdl_config(
-        train_state_metadata.input_shape_dtype, do_eval=do_eval
+        train_state_metadata.input_shape_dtype, do_eval=do_eval,
+        extra_mutable_list=DEFAULT_INIT_MUTABLE_LIST
     )
     with params_fpath.open('w') as params_file:
       hyper_params_dump = base_hyperparams.nested_struct_to_text(hyper_params)
@@ -379,7 +389,8 @@ def initialize_model_state(
     is_eval_for_init = is_eval
   if not var_weight_hparams:
     var_weight_hparams = model.abstract_init_with_metadata(
-        inputs_shape_dtype, do_eval=is_eval_for_init
+        inputs_shape_dtype, do_eval=is_eval_for_init,
+        extra_mutable_list=DEFAULT_INIT_MUTABLE_LIST
     )
   logging.info('init_var prng_seed: %s', init_key)
   logging.info('var_weight_hparams: %s', var_weight_hparams)
@@ -396,7 +407,7 @@ def initialize_model_state(
       inputs = jax.tree.map(jnp.zeros_like, inputs_shape_dtype)
       if model.hparams.fprop_dtype == jnp.bfloat16:
         inputs = jax.tree.map(_maybe_to_bfloat16, inputs)
-      return model.init(init_key, inputs)
+      return model.init(init_key, inputs, mutable=DEFAULT_INIT_MUTABLE_LIST)
 
   initial_vars = init_fn(init_key)
   logging.info('initial_vars: %s', jax.tree.map(jnp.shape, initial_vars))
@@ -809,7 +820,6 @@ class LossFnProtocol(Protocol):
   ) -> tuple[JTensor, sgf.GradAuxInfo]:
     """Produces losses and grad info by passing the inputs through a model."""
 
-
 def _get_default_loss_fn(
     jax_task: tasks_lib.SingleTask,
     context_p: base_layer.JaxContext.HParams,
@@ -920,6 +930,15 @@ class GradFnProtocol(Protocol):
       A tuple of (values, grads).
     """
 
+import gin
+
+@gin.configurable
+@dataclasses.dataclass
+class MicrobatchConfig:
+  size: Optional[int] = None
+  schedule: Optional[str] = None
+  num_stages: Optional[int] = None
+  interleave: Optional[int] = None
 
 def _get_default_grad_fn(
     excluded_for_grad: NestedMap, excluded_for_opt: NestedMap
@@ -942,11 +961,11 @@ def _get_default_grad_fn(
 
     def _loss(
         mdl_vars_grad: NestedJTensor,
-        mdl_vars_nograd_and_inputs: tuple[NestedJTensor, NestedMap],
+        mdl_vars_nograd: NestedJTensor,
+        inputs: NestedMap,
         prng_key: PRNGKey,
     ):
-      mdl_vars_nograd, inputs = mdl_vars_nograd_and_inputs
-      merged_vars = jax.tree.map(
+      merged_vars = jax.tree_map(
           lambda e, x, y: y if e else x,
           excluded_for_grad,
           mdl_vars_grad,
@@ -958,8 +977,25 @@ def _get_default_grad_fn(
       g = jax.value_and_grad(_loss, has_aux=True, allow_int=True)
     else:
       g = functools.partial(learner.stochastic_gradient.grad_fn, _loss)
-    values, grads = g(with_grad, (no_grad, inputs), prng_key)
-    grads = jax.tree.map(
+
+    microbatch_config = MicrobatchConfig()
+    if microbatch_config.size is not None:
+      from legate.jax import microbatch
+
+      def microbatch_input_sharding(x):
+        if len(x.shape) == 2:
+          return P("replica", "seq")
+        return None
+
+      arg_shardings = jax.tree_map(microbatch_input_sharding, inputs)
+
+      g = microbatch(g, dim=0, argnum=2,
+                     size=microbatch_config.size, arg_shardings=arg_shardings,
+                     schedule=microbatch_config.schedule, num_stages=microbatch_config.num_stages,
+                     interleave=microbatch_config.interleave)
+    values, grads = g(with_grad, no_grad, inputs, prng_key)
+
+    grads = jax.tree_map(
         lambda eo, eg, m, g: jnp.zeros_like(m) if eg and not eo else g,
         excluded_for_opt,
         excluded_for_grad,
@@ -994,6 +1030,7 @@ def get_excluded_var_masks(
   excluded_for_grad = tasks_lib.get_excluded_var_mask_for_grad(
       var_weight_hparams, learner
   )
+
   _log_bprop_include_exclude_list(var_weight_hparams, excluded_for_grad)
 
   # Excluded for optimizer states.
@@ -1090,7 +1127,7 @@ def train_step_single_learner(
 
   if not var_weight_hparams:
     with base_layer.JaxContext.new_context(hparams=context_p):
-      var_weight_hparams = model.abstract_init_with_metadata(inputs)
+      var_weight_hparams = model.abstract_init_with_metadata(inputs, extra_mutable_list=DEFAULT_INIT_MUTABLE_LIST)
   updated_model_vars = jax_task.maybe_adjust_train_state(  # pytype: disable=wrong-arg-types  # jax-ndarray
       step=states.step,
       mdl_vars=states.mdl_vars,
@@ -1162,6 +1199,7 @@ def train_step_single_learner(
     wps_with_opt = tasks_lib.filter_vars_for_grad_or_opt(
         var_weight_hparams, excluded_for_learner
     )
+
     transformed_grads, new_opt_states = learner.update_states(
         grads, states.opt_states[0], vars_with_opt, wps_with_opt
     )
@@ -1197,6 +1235,7 @@ def train_step_single_learner(
         states.mdl_vars,
         mdl_vars,
     )
+
     new_states = states.new_state(
         mdl_vars=mdl_vars, opt_states=[new_opt_states], extra_state=()
     )
@@ -1300,7 +1339,7 @@ def eval_step_single_learner(
     var_weight_hparams = model.abstract_init_with_metadata(
         inputs,
         do_eval=not jax_task.hparams.train.always_use_train_for_model_init,
-    )
+        extra_mutable_list=DEFAULT_INIT_MUTABLE_LIST)
 
   if fprop_dtype == jnp.float32:
     pass
@@ -1554,7 +1593,7 @@ def initialize_partitioned_model_states(
   model = jax_task.model
   if not var_weight_hparams:
     var_weight_hparams = model.abstract_init_with_metadata(
-        global_input_shapes, do_eval=is_eval
+        global_input_shapes, do_eval=is_eval, extra_mutable_list=DEFAULT_INIT_MUTABLE_LIST
     )
 
   train_state_partition_specs = (
@@ -1601,19 +1640,19 @@ def initialize_partitioned_model_states(
       lambda p: jax.sharding.NamedSharding(global_mesh, p),
       prng_key_partition_spec,
   )
-  train_state_shardings = jax.tree.map(
-      lambda p: jax.sharding.NamedSharding(global_mesh, p),
-      train_state_partition_specs,
-  )
 
-  init_fn = pjit.pjit(
-      init_model_from_seed,
-      in_shardings=prng_key_shardings,
-      out_shardings=train_state_shardings,
-  )
-  init_fn = bind_mesh(init_fn, global_mesh)
+  train_state_shardings = train_state_partition_specs
 
+  config = PaxLegateConfig()
+  with legate.jax.context(configurable=config.configurable):
+    init_fn = pjit.pjit(
+        init_model_from_seed,
+        in_shardings=prng_key_shardings,
+        out_shardings=train_state_shardings,
+    )
+  init_fn = bind_mesh(init_fn, global_mesh)
   partitioned_vars = init_fn(prng_key)
+
   train_state_provenance = train_states.build_train_state_provenance(
       partitioned_vars
   )
@@ -1647,7 +1686,9 @@ def shard_on_batch_dim_partition_spec(
   sharding = [-1] * x_dim
   # Assume the first dim is batch, and fully shard the batch dim over the entire
   # mesh.
-  sharding[0] = tuple(mesh_names)
+  sharding[0] = tuple([x for x in mesh_names if x != "seq"])
+  if x_dim > 1:
+    sharding[1] = "seq"
   return base_layer.to_partition_spec(sharding, mesh_names)
 
 
