diff --git a/praxis/base_layer.py b/praxis/base_layer.py
index 3f628b8..09fe860 100644
--- a/praxis/base_layer.py
+++ b/praxis/base_layer.py
@@ -247,7 +247,7 @@ def to_partition_spec(
               f's_i: {s_i}, mesh_axis_names: {mesh_axis_names}')
           split_dims.append(mesh_axis_names[s_i])
       elif isinstance(s_i, str):
-        assert s_i in mesh_axis_names
+        #assert s_i in mesh_axis_names
         split_dims.append(s_i)
       elif isinstance(s_i, (tuple, list)):
         split_dims.append(_parse_split_dims(s_i))
diff --git a/praxis/contrib/gpu/scripts_gpu/models.py b/praxis/contrib/gpu/scripts_gpu/models.py
index a926c58..59fd2f5 100644
--- a/praxis/contrib/gpu/scripts_gpu/models.py
+++ b/praxis/contrib/gpu/scripts_gpu/models.py
@@ -81,9 +81,9 @@ def compute_xent_loss_helper(
     )
   predicted_labels = predictions.per_example_argmax.astype(labels.dtype)
   num_preds = predictions.total_weight
-  mean_acc = jnp.sum((labels == predicted_labels) * weights) / jnp.maximum(
-      num_preds, 1
-  )
+  #mean_acc = jnp.sum((labels == predicted_labels) * weights) / jnp.maximum(
+  #    num_preds, 1
+  #)
   metric_weight = jnp.array(num_preds, predictions.avg_xent.dtype)
 
   if hasattr(predictions, 'avg_xent_weight'):
@@ -99,7 +99,7 @@ def compute_xent_loss_helper(
           jnp.array(1.0, predictions.aux_loss.dtype),
       ),
       log_pplx=(predictions.avg_xent, avg_xent_weight),
-      fraction_of_correct_next_step_preds=(mean_acc, metric_weight),
+      #fraction_of_correct_next_step_preds=(mean_acc, metric_weight),
       num_predictions=(num_preds, jnp.array(1.0, num_preds.dtype)),
   )
 
@@ -176,12 +176,12 @@ def compute_eval_metrics_helper(
     )
     num_nonpadding = jnp.sum(input_batch.eval_sample_weights)
 
-    mean_acc_strict = jnp.sum(
-        jnp.sum((labels == predicted_labels) * weights, axis=-1) == num_acc
-    ) / jnp.maximum(num_nonpadding, 1)
-    strict_weight = jnp.array(num_nonpadding, predictions.avg_xent.dtype)
-
-    metrics.acc_strict = (mean_acc_strict, strict_weight)
+    #mean_acc_strict = jnp.sum(
+    #    jnp.sum((labels == predicted_labels) * weights, axis=-1) == num_acc
+    #) / jnp.maximum(num_nonpadding, 1)
+    #strict_weight = jnp.array(num_nonpadding, predictions.avg_xent.dtype)
+    #
+    #metrics.acc_strict = (mean_acc_strict, strict_weight)
 
   elif eval_task == 'boolq':
 
diff --git a/praxis/contrib/gpu/scripts_gpu/te_helper.py b/praxis/contrib/gpu/scripts_gpu/te_helper.py
new file mode 100644
index 0000000..d3d5dbd
--- /dev/null
+++ b/praxis/contrib/gpu/scripts_gpu/te_helper.py
@@ -0,0 +1,399 @@
+import os
+
+from absl import logging
+
+
+from praxis import base_layer
+from praxis import pax_fiddle
+from praxis import pytypes
+from praxis import layers
+from praxis.layers.checkpoint_policy import AutodiffCheckpointType
+from praxis.layers import activations
+from praxis.layers import attentions, grouped_query_attention, multi_query_attention
+from praxis.layers import embedding_softmax
+from praxis.layers import normalizations
+
+from praxis.contrib.gpu.scripts_gpu.lora_layers import (
+    LoraAttentionProjection,
+    LoraCombinedQKVProjection,
+    LoraLinear,
+)
+
+
+try:
+    import transformer_engine.jax as te
+    import transformer_engine.jax.flax as te_flax
+    import transformer_engine.jax.praxis as te_praxis
+    _IS_TRANSFORMER_ENGINE_INSTALLED = True
+    import praxis.layers.repeats as praxis_repeat
+    # This is to make Repeat module correctly generate collections we need.
+    praxis_repeat.SCAN_VARIABLE_AXES.update({base_layer.NON_PAX_VAR_COLLECTION[1]: 0, # 1-idx = params_axes
+                                            te.fp8.FP8Helper.FP8_COLLECTION_NAME:0})
+    TE_PIPELINE_EXTRA_VMAP_VAR_AXES = {
+        base_layer.NON_PAX_VAR_COLLECTION[1]: 0, # 1-idx = params_axes
+        te.fp8.FP8Helper.FP8_COLLECTION_NAME:0
+    }
+
+    TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST = [te.fp8.FP8Helper.FP8_COLLECTION_NAME]
+
+    ENABLE_TE_SP = bool(int(os.environ.get('ENABLE_TE_SP', 0)))
+
+except ModuleNotFoundError as e:
+    _IS_TRANSFORMER_ENGINE_INSTALLED = False
+    TE_PIPELINE_EXTRA_VMAP_VAR_AXES = {}
+    TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST = []
+    ENABLE_TE_SP = False
+
+LayerTpl = pax_fiddle.Config[base_layer.BaseLayer]
+JTensor = pytypes.JTensor
+
+
+class TransformerEngineHelperBase:
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(fprop, deterministic):
+        raise NotImplementedError
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, layer_p, layer_id):
+        raise NotImplementedError
+
+    @staticmethod
+    def get_input_bld(original_bld, batch_axes, mdl_axis):
+        # This is used to specify the sharding pattern of inputs to TransformerLayers.
+        raise NotImplementedError
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(xformer_layer_p):
+        raise NotImplementedError
+
+    @staticmethod
+    def check_checkpoint_policy(tpl):
+        raise NotImplementedError
+
+
+class TENotInstalledHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(fprop, deterministic):
+        return fprop
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, layer_p, layer_id):
+        layer_p.name = f'layer_{layer_id}'
+        layer_p.use_cross_attention = stacked_transformer_obj.use_cross_attention
+        layer_p.num_heads = stacked_transformer_obj.num_heads
+        layer_p.dim_per_head = stacked_transformer_obj.dim_per_head
+        layer_p.input_dims = stacked_transformer_obj.model_dims
+        layer_p.packed_input = stacked_transformer_obj.packed_input
+        layer_p.atten_dropout_prob = stacked_transformer_obj.atten_dropout_prob or stacked_transformer_obj.dropout_prob
+        layer_p.residual_dropout_prob = (
+            stacked_transformer_obj.residual_dropout_prob or stacked_transformer_obj.dropout_prob
+        )
+        layer_p.relu_dropout_prob = stacked_transformer_obj.relu_dropout_prob or stacked_transformer_obj.dropout_prob
+        layer_p.hidden_dims = stacked_transformer_obj.hidden_dims
+
+        if stacked_transformer_obj.residual_droppath_prob > 0.0:
+            layer_p.residual_droppath_prob = (
+                stacked_transformer_obj.residual_droppath_prob * layer_id / max(1, stacked_transformer_obj.num_layers)
+            )
+        return layer_p
+
+    @staticmethod
+    def get_input_bld(original_bld, *_):
+        return original_bld
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(xformer_layer_p):
+        return xformer_layer_p.tr_atten_tpl.activation_split_dims_mapping.bld
+
+    @staticmethod
+    def check_checkpoint_policy(_):
+        """Every checkpoint policy is valid without TE"""
+        pass
+
+
+class TEInstalledHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(_, deterministic):
+        def _fprop(
+            transformer,
+            x_in,
+            paddings,
+            attention_mask,
+            cross_inputs,
+            cross_attention_mask,
+            segment_pos
+        ):
+            x_out = transformer(
+                inputs=x_in,
+                attention_mask=attention_mask,
+                encoded=cross_inputs,
+                encoder_decoder_mask=cross_attention_mask,
+                deterministic=deterministic)
+            return x_out
+        return _fprop
+
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, _, layer_id):
+        te_transformer_tpl = pax_fiddle.Config(te_praxis.TransformerLayer,
+            name=f'layer_{layer_id}',
+            enable_relative_embedding=False,
+            enable_sequence_parallel=ENABLE_TE_SP,
+            transpose_batch_sequence=False
+        )
+
+        def update_ln_te_tpl(te_tpl, transformer_layer_tpl):
+            # TE requires all normalization are the same
+            assert transformer_layer_tpl.ln_tpl == transformer_layer_tpl.tr_fflayer_tpl.ln_tpl
+            ln_tpl = transformer_layer_tpl.ln_tpl
+            if issubclass(ln_tpl.cls, normalizations.LayerNorm):
+                te_tpl.layernorm_type = 'layernorm'
+                assert ln_tpl.use_scale
+                assert ln_tpl.use_bias
+            elif issubclass(ln_tpl.cls, normalizations.RmsNorm):
+                te_tpl.layernorm_type = 'rmsnorm'
+            else:
+                raise ValueError(f'Unsupported {ln_tpl.cls=}, LayerNorm, RmsNorm are supported.')
+            te_tpl.zero_centered_gamma = not ln_tpl.direct_scale
+            te_tpl.layernorm_epsilon = ln_tpl.epsilon
+            return te_tpl
+
+        def update_ff_te_tpl(te_tpl, ff_layer_tpl):
+            mlp_activations = ()
+            if ff_layer_tpl.use_gated_activation:
+               mlp_activations += ('linear',)
+
+            if issubclass(ff_layer_tpl.activation_tpl.cls, activations.Identity):
+                mlp_activations += ('linear',)
+            else:
+                mlp_activations += (ff_layer_tpl.activation_tpl.cls.__name__.lower(),)
+
+            te_tpl.mlp_activations = mlp_activations
+            return te_tpl
+
+        def update_attn_te_tpl(te_tpl, attn_tpl):
+            if issubclass(attn_tpl.cls, attentions.DotProductAttention):
+                # Check the DotProductAttention parameters are aligned to TE's attention
+                assert attn_tpl.internal_enable_query_scale or attn_tpl.scale_logits_by_head_dims
+                assert not (attn_tpl.internal_enable_query_scale and attn_tpl.scale_logits_by_head_dims)
+                assert not attn_tpl.internal_enable_per_dim_scale
+                assert not attn_tpl.scale_query_by_dim_per_head
+                assert not attn_tpl.dconv_qkv
+                assert not attn_tpl.internal_gshard_gaussian_init
+                assert attn_tpl.relative_bias_tpl is None
+                assert attn_tpl.attention_extra_logit is None
+                assert attn_tpl.ngrammer_tpl is None
+                te_tpl.enable_rotary_pos_emb = attn_tpl.use_rotary_position_emb
+                if attn_tpl.rotary_position_emb_tpl.cls == embedding_softmax.RotaryPositionalEmbedding:
+                    te_tpl.rotary_pos_emb_group_method = 'alternate'
+            elif issubclass(attn_tpl.cls, grouped_query_attention.GroupedQueryAttention):
+                te_tpl.num_gqa_groups = attn_tpl.num_kv_heads
+                if attn_tpl.rope_min_max_timescales is not None:
+                    te_tpl.enable_rotary_pos_emb = True
+                    te_tpl.rotary_pos_emb_windows = attn_tpl.rope_min_max_timescales
+                assert attn_tpl.atten_temp == 1.
+            elif issubclass(attn_tpl.cls, multi_query_attention.MultiQueryDotProductAttention):
+                te_tpl.num_gqa_groups = attn_tpl.num_kv_heads
+                te_tpl.enable_rotary_pos_emb = attn_tpl.use_rotary_position_emb
+                if attn_tpl.rotary_position_emb_tpl.cls == embedding_softmax.RotaryPositionalEmbedding:
+                    te_tpl.rotary_pos_emb_group_method = 'alternate'
+            else:
+                raise ValueError(f'Unsupported {attn_tpl.cls=}')
+            assert attn_tpl.atten_logit_cap <= 0., 'atten_logit_cap > 0. is not supported in TE'
+            te_tpl.scaled_query_init = False
+            te_tpl.scale_attn_logits = True
+            return te_tpl
+
+        transformer_layer_tpl = stacked_transformer_obj.transformer_layer_params_tpl
+        # Update TE normalization layer configs
+        te_transformer_tpl = update_ln_te_tpl(te_transformer_tpl, transformer_layer_tpl)
+        # Update TE feed forward layer configs
+        te_transformer_tpl = update_ff_te_tpl(te_transformer_tpl, transformer_layer_tpl.tr_fflayer_tpl)
+        # Update TE attention layer configs
+        te_transformer_tpl = update_attn_te_tpl(te_transformer_tpl, transformer_layer_tpl.tr_atten_tpl)
+        # TE currently only allow the bias config to be same between feed forward, qkv proj, out proj
+        assert (transformer_layer_tpl.tr_fflayer_tpl.has_bias ==
+            transformer_layer_tpl.tr_atten_tpl.use_bias), "TE only allows same bias settings."
+        te_transformer_tpl.use_bias = transformer_layer_tpl.tr_fflayer_tpl.has_bias
+        te_transformer_tpl.self_attn_mask_type = 'padding_causal' \
+            if stacked_transformer_obj.mask_self_attention else 'padding'
+
+        te_transformer_tpl.logical_axes_rules = te_flax.extend_logical_axis_rules(tuple())
+        te_transformer_tpl.params_init = stacked_transformer_obj.params_init
+        te_transformer_tpl.dtype = stacked_transformer_obj.fprop_dtype
+        te_transformer_tpl.layer_type = te_praxis.TransformerLayerType.DECODER if stacked_transformer_obj.use_cross_attention \
+                        else te_praxis.TransformerLayerType.ENCODER
+        te_transformer_tpl.num_attention_heads = stacked_transformer_obj.num_heads
+        te_transformer_tpl.hidden_size = stacked_transformer_obj.model_dims
+        te_transformer_tpl.mlp_hidden_size = stacked_transformer_obj.hidden_dims
+        te_transformer_tpl.dropout_rng_name = base_layer.RANDOM
+        te_transformer_tpl.attention_dropout = stacked_transformer_obj.atten_dropout_prob or stacked_transformer_obj.dropout_prob
+        te_transformer_tpl.hidden_dropout = stacked_transformer_obj.residual_dropout_prob or stacked_transformer_obj.dropout_prob
+        te_transformer_tpl.intermediate_dropout = stacked_transformer_obj.relu_dropout_prob or stacked_transformer_obj.dropout_prob
+        if stacked_transformer_obj.residual_droppath_prob > 0.0:
+            te_transformer_tpl.drop_path = (
+                stacked_transformer_obj.residual_droppath_prob * layer_id / max(1, stacked_transformer_obj.num_layers)
+            )
+
+        assert stacked_transformer_obj.dim_per_head == stacked_transformer_obj.model_dims // stacked_transformer_obj.num_heads
+        assert stacked_transformer_obj.packed_input == False
+        assert len(stacked_transformer_obj.moe_layers) == 0
+        assert stacked_transformer_obj.ngrammer_tpls is None
+
+        def update_lora_te_tpl(te_tpl, transformer_layer_tpl):
+            lora_enabled = False
+            te_lora_scope = "none"
+            lora_rank = None
+            if (
+                transformer_layer_tpl.tr_fflayer_tpl.fflayer_tpl.linear_tpl.__fn_or_cls__
+                is LoraLinear
+            ):
+                lora_enabled = True
+                mlp_included_in_lora = True
+                current_rank = (
+                    transformer_layer_tpl.tr_fflayer_tpl.fflayer_tpl.linear_tpl.rank
+                )
+                lora_rank = (
+                    current_rank if lora_rank is None else lora_rank & current_rank
+                )
+
+            attention_included_in_lora = False
+            if (
+                hasattr(transformer_layer_tpl.tr_atten_tpl, "combined_qkv_proj_tpl")
+                and transformer_layer_tpl.tr_atten_tpl.combined_qkv_proj_tpl.__fn_or_cls__
+                is LoraCombinedQKVProjection
+            ):
+                lora_enabled = True
+                attention_included_in_lora = True
+                current_rank = (
+                    transformer_layer_tpl.tr_atten_tpl.combined_qkv_proj_tpl.rank
+                )
+                lora_rank = (
+                    current_rank if lora_rank is None else lora_rank & current_rank
+                )
+
+            if (
+                hasattr(transformer_layer_tpl.tr_atten_tpl, "proj_tpl")
+                and transformer_layer_tpl.tr_atten_tpl.proj_tpl.__fn_or_cls__
+                is LoraAttentionProjection
+            ):
+                lora_enabled = True
+                attention_included_in_lora = True
+                current_rank = transformer_layer_tpl.tr_atten_tpl.proj_tpl.rank
+                lora_rank = (
+                    current_rank if lora_rank is None else lora_rank & current_rank
+                )
+
+            if lora_enabled:
+                assert (
+                    lora_rank > 0
+                ), "LoRA rank should be the same for all layers and greater than 0."
+                if attention_included_in_lora and mlp_included_in_lora:
+                    te_lora_scope = "all"
+                elif attention_included_in_lora and not mlp_included_in_lora:
+                    te_lora_scope = "exclude_mlp"
+                elif mlp_included_in_lora and not attention_included_in_lora:
+                    te_lora_scope = "mlp"
+
+                te_transformer_tpl.low_rank_adaptation_scope = te_lora_scope
+                te_transformer_tpl.low_rank_adaptation_dim = lora_rank
+
+            return te_tpl
+
+        try:
+            te_transformer_tpl = update_lora_te_tpl(
+                te_transformer_tpl, transformer_layer_tpl
+            )
+        except Exception as e:
+            logging.warning(f"Unable to use LoRA with TE: {e}")
+        
+
+        return te_transformer_tpl
+
+    @staticmethod
+    def get_input_bld(original_bld, batch_axes, mdl_axis):
+        if ENABLE_TE_SP:
+            return [batch_axes, mdl_axis, None]
+        return original_bld
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(_):
+        rules = te_flax.extend_logical_axis_rules(tuple())
+        # rules [(batch_axis_name, ('replicat', 'data'))', ...)]
+        batch_mapping = rules[0][1]
+        hidden_tp_mapping = rules[4][1]
+        # [Batch, Seqlen, Hidden]
+        bld_mapping = [batch_mapping, None, hidden_tp_mapping]
+        return bld_mapping
+
+    @staticmethod
+    def check_checkpoint_policy(tpl):
+        """Some checkpoint policies are not compatible with TE fused attention"""
+        if issubclass(tpl.cls, layers.transformers.StackedTransformer):
+            remat = tpl.remat
+            attention_dropout = tpl.atten_dropout_prob or tpl.dropout_prob
+        elif issubclass(tpl.cls, layers.transformers.StackedTransformerRepeated):
+            if not issubclass(tpl.block.cls, layers.transformers.StackedTransformer):
+                return
+            remat = True  # Current StackedTransformerRepeat always enables remat
+            attention_dropout = tpl.block.atten_dropout_prob or tpl.block.dropout_prob
+        else:
+            raise ValueError(f'Unsupported class={tpl.cls}')
+
+        supported_checkpoint_policies = [
+            AutodiffCheckpointType.SAVE_CONTEXT,
+            AutodiffCheckpointType.SAVE_CONTEXT_AND_OUT_PROJ,
+            AutodiffCheckpointType.SAVE_DOT_FOR_MLPERF_200B,
+            AutodiffCheckpointType.SAVE_QUANTIZED,
+            AutodiffCheckpointType.SAVE_DOT_EXCEPT_LOGITS_FFN1,
+            AutodiffCheckpointType.SAVE_DOT_EXCEPT_LOGITS]
+        fused_attn_enabled = int(os.getenv("NVTE_FUSED_ATTN", "0"))
+        if remat and fused_attn_enabled and attention_dropout > 0.:
+            assert tpl.checkpoint_policy in supported_checkpoint_policies, \
+            "Fused attn in TE only permits policies that save 'context' tensors when dropout is " \
+            "enabled. This restriction is due to the maintenance of the dropout offset within TE, " \
+            "which is incompatible with the JAX remat. Consequently, it's necessary to bypass " \
+            "recomputation in the attention layer when fused attention is activated. The supported " \
+            f"checkpoint_policies are {supported_checkpoint_policies} but the provided " \
+            f"checkpoint_policy is '{tpl.checkpoint_policy}'."
+
+
+class TransformerEngineHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def is_enabled_te():
+        enable_te = bool(int((os.environ.get("ENABLE_TE", False))))
+        return (_IS_TRANSFORMER_ENGINE_INSTALLED and enable_te)
+
+    @staticmethod
+    def get_helper():
+        if TransformerEngineHelper.is_enabled_te():
+            return TEInstalledHelper
+        return TENotInstalledHelper
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(fprop, deterministic):
+        return TransformerEngineHelper.get_helper().get_fprop_caller_of_stack_transformer(
+                    fprop, deterministic)
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, layer_p, layer_id):
+        return TransformerEngineHelper.get_helper().set_layer_params_to_stack_transformer(
+                    stacked_transformer_obj, layer_p, layer_id)
+
+    @staticmethod
+    def get_input_bld(original_bld, batch_axes, mdl_axis):
+        return TransformerEngineHelper.get_helper().get_input_bld(
+                    original_bld, batch_axes, mdl_axis)
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(xformer_layer_p):
+        return TransformerEngineHelper.get_helper().get_bld_mapping_for_pipelined_transformer(
+                    xformer_layer_p)
+
+    @staticmethod
+    def check_checkpoint_policy(tpl):
+        return TransformerEngineHelper.get_helper().check_checkpoint_policy(tpl)
diff --git a/praxis/layers/attentions.py b/praxis/layers/attentions.py
index ed9bc9f..fb36427 100644
--- a/praxis/layers/attentions.py
+++ b/praxis/layers/attentions.py
@@ -1182,6 +1182,7 @@ class DotProductAttention(base_layer.BaseLayer):
   decode_cache: bool = True
   attention_mask_summary: bool = False
   zero_fully_masked: bool = False
+  mha_mask_addition_pattern: bool = True
   qk_einsum_tpl: LayerTpl = template_field(base_ops.EinsumOp)
   pv_einsum_tpl: LayerTpl = template_field(base_ops.EinsumOp)
   per_dim_scale_tpl: LayerTpl = template_field(PerDimScale)
@@ -1541,7 +1542,11 @@ class DotProductAttention(base_layer.BaseLayer):
     # Attention softmax is always carried out in fp32.
     logits = logits.astype(jnp.float32)
     # Apply attention masking
-    padded_logits = py_utils.apply_mask_to_logits(logits, atten_mask)
+    if self.mha_mask_addition_pattern:
+      padded_logits = logits + atten_mask.astype(jnp.float32)
+    else:
+      padded_logits = py_utils.apply_mask_to_logits(logits, atten_mask)
+
     if self.attention_mask_summary:
       self.add_summary('attention_mask', atten_mask)
     if self.attention_extra_logit is None:
diff --git a/praxis/layers/checkpoint_policy.py b/praxis/layers/checkpoint_policy.py
index 5f173fd..3566142 100644
--- a/praxis/layers/checkpoint_policy.py
+++ b/praxis/layers/checkpoint_policy.py
@@ -40,6 +40,7 @@ class AutodiffCheckpointType(str, enum.Enum):
   SAVE_QKV_OUT_PROJ_SEPARATE = 'save_qkv_out_proj_separate'
   SAVE_DOT_EXCEPT_LOGITS_FFN1 = 'save_dot_except_logits_ffn1'
   SAVE_DOT_EXCEPT_LOGITS = 'save_dot_except_logits'
+  SAVE_QKV_OUT_PROJ_FFN = 'save_qkv_out_proj_ffn'
 
 
 def custom_policy(checkpoint_policy: AutodiffCheckpointType):
@@ -52,6 +53,10 @@ def custom_policy(checkpoint_policy: AutodiffCheckpointType):
     return jax.checkpoint_policies.checkpoint_dots
   if checkpoint_policy == AutodiffCheckpointType.SAVE_DOT_WITH_NO_BATCH_DIM:
     return jax.checkpoint_policies.checkpoint_dots_with_no_batch_dims
+  if checkpoint_policy == AutodiffCheckpointType.SAVE_QKV_OUT_PROJ_FFN:
+    return jax.checkpoint_policies.save_only_these_names(
+        'combined_qkv_proj', 'out_proj', 'ffn1', 'ffn2'
+    )
   if checkpoint_policy == AutodiffCheckpointType.SAVE_QKV_OUT_PROJ:
     return jax.checkpoint_policies.save_only_these_names(
         'combined_qkv_proj', 'out_proj'
diff --git a/praxis/layers/embedding_softmax.py b/praxis/layers/embedding_softmax.py
index c41ef48..08160ae 100644
--- a/praxis/layers/embedding_softmax.py
+++ b/praxis/layers/embedding_softmax.py
@@ -320,14 +320,22 @@ class FullSoftmax(base_layer.BaseLayer):
       )
       logits = log_probs = None
 
-    # Compute total softmax cross-entropy loss for the output tensor.
-    total_xent = jnp.sum(
+    per_batch_xent = jnp.sum(
         jnp.expand_dims(per_example_xent, axis=-1) * class_weights,
         dtype=jnp.float32,
+        axis=[-2,-1],
+    )
+    per_batch_weight = jnp.sum(
+        class_weights, axis=[-2,-1], dtype=jnp.float32
     )
-    total_weight = jnp.sum(class_weights, dtype=jnp.float32)
+
+    avg_batch_xent = per_batch_xent / per_batch_weight
+    avg_xent = jnp.sum(avg_batch_xent)
+    total_xent = jnp.sum(per_batch_xent)
+    total_weight = jnp.sum(per_batch_weight)
 
     if self.z_loss_weight > 0.0:
+      raise Exception("z_loss_weight not supported")
       assert logits is not None
       z_loss = (
           jnp.sum(_compute_z_loss(logits) * class_weights, dtype=jnp.float32)
@@ -348,7 +356,7 @@ class FullSoftmax(base_layer.BaseLayer):
         per_example_xent=per_example_xent.astype(jnp.float32),
         total_xent=total_xent,
         total_weight=total_weight,
-        avg_xent=(total_xent / (total_weight + 1e-6)).astype(jnp.float32),
+        avg_xent=avg_xent,
     )
     if class_ids is not None:
       output_nmap.accuracy = (
diff --git a/praxis/layers/grok.py b/praxis/layers/grok.py
index f66d143..d1effc4 100644
--- a/praxis/layers/grok.py
+++ b/praxis/layers/grok.py
@@ -59,6 +59,7 @@ def GrokStackedTransformerHParams(
     combine_qkv=False,
     bidirectional=False,
     use_fp8=False,
+    use_te_dpa=False,
 ) -> pax_fiddle.Config[transformers.StackedTransformer]:
   """Common setup for Grok-1 Transformer layers.
 
@@ -168,6 +169,7 @@ def GrokStackedTransformerHParams(
   p.transformer_layer_params_tpl.tr_atten_tpl = pax_fiddle.Config(
       multi_query_attention.MultiQueryDotProductAttention,
       num_kv_heads=attention_num_groups,
+      use_te_dpa=use_te_dpa,
   )
   tr_atten_tpl = p.transformer_layer_params_tpl.tr_atten_tpl
   tr_atten_tpl.combine_qkv = False
@@ -225,6 +227,7 @@ def GrokUniTransformerLmHParams(
     model_type=LanguageModelType.CAUSAL,
     checkpoint_policy=AutodiffCheckpointType.SAVE_NOTHING,
     use_fp8=False,
+    use_te_dpa=False,
 ) -> pax_fiddle.Config[transformer_models.TransformerLm]:
   """Common setup for Grok-1 Decoder-only Transformer Model.
 
@@ -328,6 +331,7 @@ def GrokUniTransformerLmHParams(
       bidirectional=bidirectional,
       moe_gating_embedding_level=moe_gating_embedding_level,
       use_fp8=use_fp8,
+      use_te_dpa=use_te_dpa,
   )
   num_blocks = num_transformer_layers
 
diff --git a/praxis/layers/multi_query_attention.py b/praxis/layers/multi_query_attention.py
index 67bdf55..67d3557 100644
--- a/praxis/layers/multi_query_attention.py
+++ b/praxis/layers/multi_query_attention.py
@@ -17,7 +17,7 @@
 
 import math
 from typing import Callable, Mapping, Sequence
-
+from absl import logging
 from flax import linen as nn
 import jax
 from jax import numpy as jnp
@@ -31,7 +31,7 @@ from praxis.layers import attentions
 from praxis.layers import base_ops
 from praxis.layers import embedding_softmax
 from praxis.layers import stochastics
-
+from transformer_engine.jax.flax.transformer import DotProductAttention as TEDotProductAttention
 
 WeightInit = base_layer.WeightInit
 WeightHParams = base_layer.WeightHParams
@@ -209,6 +209,7 @@ class MultiQueryDotProductAttention(base_layer.BaseLayer):
   pv_einsum_tpl: LayerTpl = template_field(base_ops.EinsumOp)
   scale_query_by_dim_per_head: bool = False
   chunked_attn_num_seq_split: int = 1
+  use_te_dpa: bool = False  # Experimental way to use TE flash attention when can't use standard TE
 
   # SPMD partition related params.
   #
@@ -347,6 +348,20 @@ class MultiQueryDotProductAttention(base_layer.BaseLayer):
     self.create_child('post', post_proj_p)
     self.create_child('qk_einsum', self.qk_einsum_tpl.clone())
     self.create_child('pv_einsum', self.pv_einsum_tpl.clone())
+    self.dpa_layer = TEDotProductAttention(
+                  head_dim=dim_per_head,
+                  num_attention_heads=self.num_heads,
+                  num_gqa_groups=self.num_kv_heads,
+                  attn_mask_type='causal', # 'causal' or 'padding'
+                  attn_bias_type='no_bias', # 'no_bias', 'pre_scale_bias' or 'post_scale_bias'
+                  attention_dropout=0.,
+                  dropout_rng_name='aqt',
+                  dtype=jnp.bfloat16,
+                  float32_logits=False,
+                  qkv_layout='BSHD_BSHD_BSHD', # 'BS3HD', 'BSHD_BS2HD' or 'BSHD_BSHD_BSHD'
+                  scale_factor=1.0/math.sqrt(self.num_heads),
+                  transpose_batch_sequence=False
+            ) 
 
   def _shard_bnh(self, x: JTensor) -> JTensor:
     """Shards tensors of shape [b, n, h].
@@ -828,29 +843,36 @@ class MultiQueryDotProductAttention(base_layer.BaseLayer):
     else:
       key_proj = self._shard_blnh(key_proj)
       value_proj = self._shard_blnh(value_proj)
-      b, t, n, h = query_proj.shape
-      _, s, nk, _ = key_proj.shape
-      assert n % nk == 0
-      v_q = jnp.reshape(query_proj, (b, t, nk, n // nk, h))
-      if relative_bias is not None:
-        v_rb = jnp.reshape(relative_bias, (b, nk, n // nk, t, s))
-      else:
-        v_rb = None
-      with self._context_for_kv_vmap():
-        encoded, atten_probs = jax.vmap(
-            self._dot_atten,
-            in_axes=(2, 2, 2, None, 1),
-            out_axes=(2, 1),
-        )(
-            v_q,
-            key_proj,
-            value_proj,
-            atten_mask,
-            v_rb,
+      if self.use_te_dpa:
+        logging.warning(
+            'use_te_dpa is set to True, so TE dpa is used as an experimental way to use TE flash attention.'
         )
-      encoded = self._shard_blnh(jnp.reshape(encoded, (b, t, n, h)))
-      if atten_probs is not None:
-        atten_probs = jnp.reshape(atten_probs, (b, t, n, s))
+        atten_probs = None
+        encoded = self.dpa_layer(query_proj, key_proj, value_proj)
+      else:
+        b, t, n, h = query_proj.shape
+        _, s, nk, _ = key_proj.shape
+        assert n % nk == 0
+        v_q = jnp.reshape(query_proj, (b, t, nk, n // nk, h))
+        if relative_bias is not None:
+          v_rb = jnp.reshape(relative_bias, (b, nk, n // nk, t, s))
+        else:
+          v_rb = None
+        with self._context_for_kv_vmap():
+          encoded, atten_probs = jax.vmap(
+              self._dot_atten,
+              in_axes=(2, 2, 2, None, 1),
+              out_axes=(2, 1),
+          )(
+              v_q,
+              key_proj,
+              value_proj,
+              atten_mask,
+              v_rb,
+          )
+        encoded = self._shard_blnh(jnp.reshape(encoded, (b, t, n, h)))
+        if atten_probs is not None:
+          atten_probs = jnp.reshape(atten_probs, (b, t, n, s))
 
     # Post projection
     encoded = self.post(encoded)
diff --git a/praxis/layers/pipeline.py b/praxis/layers/pipeline.py
index fb2f72f..56f8787 100644
--- a/praxis/layers/pipeline.py
+++ b/praxis/layers/pipeline.py
@@ -28,6 +28,8 @@ from praxis import pax_fiddle
 from praxis import py_utils
 from praxis import pytypes
 from praxis.layers import checkpoint_policy
+from praxis.contrib.gpu.scripts_gpu.te_helper import TE_PIPELINE_EXTRA_VMAP_VAR_AXES
+from praxis.contrib.gpu.scripts_gpu.te_helper import TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST
 
 NestedMap = py_utils.NestedMap
 JTensor = pytypes.JTensor
@@ -420,6 +422,7 @@ class LayerwiseShardablePipelined(base_layer.BaseLayer):
             NON_TRAINABLE: 0,
             INTERMEDIATES: 0,
             HYPER_PARAMS: 0,
+            **TE_PIPELINE_EXTRA_VMAP_VAR_AXES
         },
         split_rngs={PARAMS: self.is_initializing(), RANDOM: True},
         metadata_params={
@@ -840,7 +843,7 @@ class LayerwiseShardablePipelined(base_layer.BaseLayer):
     #
     # Note that fprop should not use PARAMS rng because there is no var init.
     variable_carry = []
-    variable_broadcast = [PARAMS]
+    variable_broadcast = [PARAMS] + TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST
     if self.is_mutable_collection(NON_TRAINABLE):
       variable_carry.append(NON_TRAINABLE)
     else:
@@ -863,7 +866,7 @@ class LayerwiseShardablePipelined(base_layer.BaseLayer):
     if bf16_vars_to_convert is not None:
       scan_fn = nn.map_variables(
           scan_fn,
-          mapped_collections=[PARAMS],
+          mapped_collections=[PARAMS, 'fp8_meta_collection'],
           mutable=True,
           trans_in_fn=_get_to_f32_converter(bf16_vars_to_convert),
           trans_out_fn=_get_to_bf16_converter(bf16_vars_to_convert),
diff --git a/praxis/layers/transformer_models.py b/praxis/layers/transformer_models.py
index 8709b27..09c8786 100644
--- a/praxis/layers/transformer_models.py
+++ b/praxis/layers/transformer_models.py
@@ -33,6 +33,7 @@ from praxis.layers import embedding_softmax
 from praxis.layers import multi_query_attention
 from praxis.layers import normalizations
 from praxis.layers import transformers
+from praxis.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper
 
 NestedMap = py_utils.NestedMap
 JTensor = pytypes.JTensor
@@ -539,16 +540,15 @@ class TransformerLm(base_layer.BaseLayer):
         if training_optimized
         else [batch_axes, None, None]
     )
+    bld = TransformerEngineHelper.get_input_bld(bld, batch_axes, mdl_axis)
+
     egcm = (
         [data_axis, None, None, mdl_axis]
         if training_optimized
         else [batch_axes, None, None, None]
     )
 
-    if seq_axis is None:
-      w_data_axes = data_axis
-    else:
-      w_data_axes = (data_axis, seq_axis)
+    w_data_axes = data_axis
 
     # w_df: sharding for weight of ffn0, shape (d, f). ff1 weights will be
     # inferred from it.
diff --git a/praxis/layers/transformers.py b/praxis/layers/transformers.py
index 74ed8b8..6aceb45 100644
--- a/praxis/layers/transformers.py
+++ b/praxis/layers/transformers.py
@@ -40,6 +40,7 @@ from praxis.layers import pipeline
 from praxis.layers import repeats
 from praxis.layers import stats
 from praxis.layers import stochastics
+from praxis.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper
 
 NestedMap = py_utils.NestedMap
 WeightInit = base_layer.WeightInit
@@ -1375,12 +1376,15 @@ class Transformer(base_layer.BaseLayer):
       atten_probs: A NestedMap with keys `self_atten` <float>[B, N, T, T].
     """
 
+    """
     inputs_stats = stats.compute_stats(inputs, jnp.expand_dims(paddings, -1))
     self.add_summary('xformer_input_mean', inputs_stats.mean_v, verbosity=3)
     self.add_summary('xformer_input_std', inputs_stats.std_v, verbosity=3)
     self.add_summary('xformer_input_abs_max', inputs_stats.max_v, verbosity=3)
 
     self.add_summary('attention_input_rms', _rms(inputs), verbosity=4)
+    """
+    
     if self.norm_policy == 'primer_hybrid':
       inputs_normalized = self.pre_layer_norm(inputs)
     elif self.norm_policy == 'pre':
@@ -1391,7 +1395,7 @@ class Transformer(base_layer.BaseLayer):
     # Compute self-attention, key/value vectors are the input itself
     atten_output, self_atten_probs = self.self_attention(
         inputs_normalized,
-        inputs_normalized,
+         inputs_normalized,
         inputs_normalized,
         atten_mask=attention_mask,
         query_segment_pos=segment_pos,
@@ -1726,23 +1730,10 @@ class StackedTransformer(base_layer.BaseLayer):
         p_i = self._clone_layer_params(self.transformer_layer_params_tpl[ii])
       else:
         p_i = self._clone_layer_params(self.transformer_layer_params_tpl)
-      p_i.name = f'layer_{i}'
-      p_i.use_cross_attention = self.use_cross_attention
-      p_i.num_heads = self.num_heads
-      p_i.dim_per_head = self.dim_per_head
-      p_i.input_dims = self.model_dims
-      p_i.packed_input = self.packed_input
-      p_i.atten_dropout_prob = self.atten_dropout_prob or self.dropout_prob
-      p_i.residual_dropout_prob = (
-          self.residual_dropout_prob or self.dropout_prob
-      )
-      p_i.relu_dropout_prob = self.relu_dropout_prob or self.dropout_prob
-      p_i.hidden_dims = self.hidden_dims
 
-      if self.residual_droppath_prob > 0.0:
-        p_i.residual_droppath_prob = (
-            self.residual_droppath_prob * i / max(1, self.num_layers)
-        )
+      TransformerEngineHelper.check_checkpoint_policy(self._to_fdl_config())
+
+      p_i = TransformerEngineHelper.set_layer_params_to_stack_transformer(self, p_i, i)
 
       if self.moe_layers and i in self.moe_layers:
         assert self.num_experts > 0
@@ -1861,6 +1852,8 @@ class StackedTransformer(base_layer.BaseLayer):
       )
       return x_out
 
+    _fprop = TransformerEngineHelper.get_fprop_caller_of_stack_transformer(_fprop, self.do_eval)
+
     fprop = _fprop
     if self.remat:
       fprop = nn.remat(
@@ -2053,6 +2046,8 @@ class StackedTransformerRepeated(base_layer.BaseLayer):
   def setup(self) -> None:
     wp = self.weight_split_dims_mapping
 
+    TransformerEngineHelper.check_checkpoint_policy(self._to_fdl_config())
+
     repeat_l_params = pax_fiddle.Config(
         repeats.Repeat,
         sub_tpl=self.block,
@@ -2330,7 +2325,7 @@ class PipelinedTransformer(base_layer.BaseLayer):
     else:
       assert self.pipeline_stage.cls == StackedTransformerRepeated
       xformer_layer_p = self.pipeline_stage.block.transformer_layer_params_tpl
-    bld_mapping = xformer_layer_p.tr_atten_tpl.activation_split_dims_mapping.bld
+    bld_mapping = TransformerEngineHelper.get_bld_mapping_for_pipelined_transformer(xformer_layer_p)
     if not self.stream_io:
       # Annotate the inputs before the pipeline to prevent unexpected
       # propagation from earlier layers.
diff --git a/praxis/py_utils.py b/praxis/py_utils.py
index 88b7bea..f731391 100644
--- a/praxis/py_utils.py
+++ b/praxis/py_utils.py
@@ -527,6 +527,7 @@ def maybe_pad_uneven_sharding(
   """Pads xs to make them evenly shardable, if needed."""
 
   def _maybe_pad(x, pspec, shape):
+    return x
     if is_optax_masked_node(x):
       return x
     paddings = get_uneven_sharding_paddings(pspec, shape, mesh_shape,
diff --git a/requirements.in b/requirements.in
index c277921..66260d4 100644
--- a/requirements.in
+++ b/requirements.in
@@ -8,8 +8,8 @@ clu @ git+https://github.com/google/CommonLoopUtils
 einops
 etils
 fiddle @ git+https://github.com/google/fiddle
-flax @ git+https://github.com/google/flax
-jax @ git+https://github.com/google/jax
+flax
+jax
 jax-bitempered-loss
 jaxtyping
 lingvo
