diff --git a/transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu b/transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu
index 42fb779..4592d85 100644
--- a/transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu
+++ b/transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu
@@ -122,7 +122,7 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
                    std::shared_ptr<fe::graph::Tensor_attributes>>;  // dropout_offset
 
     using CacheType = std::map<FADescriptor_v1, graph_and_tensors>;
-    static thread_local CacheType sdpa_f16_fprop_cache;
+    static CacheType sdpa_f16_fprop_cache;
 
     // Get plan from cache if cache is available, otherwise create one
     auto get_graph = [&](CacheType &cache, const FADescriptor_v1 &descriptor) -> graph_and_tensors {
@@ -461,7 +461,7 @@ void fused_attn_arbitrary_seqlen_bwd_impl(
                    std::shared_ptr<fe::graph::Tensor_attributes>>;  // dropout_offset
 
     using CacheType = std::map<FADescriptor_v1, graph_and_tensors>;
-    static thread_local CacheType sdpa_f16_bprop_cache;
+    static CacheType sdpa_f16_bprop_cache;
 
     // Get plan from cache if cache is available, otherwise create one
     auto get_graph = [&](CacheType &cache, const FADescriptor_v1 &descriptor) -> graph_and_tensors {
diff --git a/transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.cu b/transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.cu
index 88c1490..cd8cb66 100644
--- a/transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.cu
+++ b/transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.cu
@@ -635,7 +635,7 @@ void fused_attn_max_512_fwd_impl(
                             false};
 
     using CacheType = std::map<FADescriptor, cudnn_frontend::ExecutionPlan>;
-    static thread_local CacheType fmha_fprop_cache;
+    static CacheType fmha_fprop_cache;
 
     // softmax auxiliary is only used in the training mode
     bool enable_dropout = is_training && (dropout_probability != 0.0f);
@@ -834,7 +834,7 @@ void fused_attn_max_512_bwd_impl(int64_t b, int64_t h, int64_t s_q, int64_t s_kv
         layout, bias_type, mask_type, tensorType, false};
 
     using CacheType = std::map<FADescriptor, cudnn_frontend::ExecutionPlan>;
-    static thread_local CacheType fmha_bprop_cache;
+    static CacheType fmha_bprop_cache;
 
     auto get_plan = [&](CacheType &cache, const FADescriptor &descriptor) {
       auto it = cache.find(descriptor);
diff --git a/transformer_engine/common/fused_attn/fused_attn_fp8.cu b/transformer_engine/common/fused_attn/fused_attn_fp8.cu
index fb7765e..371dd3c 100644
--- a/transformer_engine/common/fused_attn/fused_attn_fp8.cu
+++ b/transformer_engine/common/fused_attn/fused_attn_fp8.cu
@@ -918,7 +918,7 @@ void fused_attn_fp8_fwd_impl(int64_t b, int64_t h, int64_t s_q, int64_t s_kv, in
                             false};
 
     using CacheType = std::map<FADescriptor, cudnn_frontend::ExecutionPlan>;
-    static thread_local CacheType fa_fprop_cache;
+    static CacheType fa_fprop_cache;
 
     // Get plan from cache if cache is available, otherwise create one
     auto get_plan = [&](CacheType& cache, const FADescriptor& descriptor) {
@@ -1199,7 +1199,7 @@ void fused_attn_fp8_bwd_impl(
                             false};
 
     using CacheType = std::map<FADescriptor, cudnn_frontend::ExecutionPlan>;
-    static thread_local CacheType fa_bprop_cache;
+    static CacheType fa_bprop_cache;
 
     // Get plan from cache if cache is available, otherwise create one
     auto get_plan = [&](CacheType& cache, const FADescriptor& descriptor) {
@@ -1718,7 +1718,7 @@ void fused_attn_fp8_fwd_impl_v1(
                    std::shared_ptr<fe::graph::Tensor_attributes>>;  // dropout_offset
 
     using CacheType = std::map<FADescriptor_v1, graph_and_tensors>;
-    static thread_local CacheType sdpa_fp8_fprop_cache;
+    static CacheType sdpa_fp8_fprop_cache;
 
     // Get plan from cache if cache is available, otherwise create one
     auto get_graph = [&](CacheType& cache, const FADescriptor_v1& descriptor) -> graph_and_tensors {
@@ -2035,7 +2035,7 @@ void fused_attn_fp8_bwd_impl_v1(
                    std::shared_ptr<fe::graph::Tensor_attributes>>;  // dropout_offset
 
     using CacheType = std::map<FADescriptor_v1, graph_and_tensors>;
-    static thread_local CacheType sdpa_fp8_bprop_cache;
+    static CacheType sdpa_fp8_bprop_cache;
 
     // Get plan from cache if cache is available, otherwise create one
     auto get_graph = [&](CacheType& cache, const FADescriptor_v1& descriptor) -> graph_and_tensors {
diff --git a/transformer_engine/common/fused_attn/utils.h b/transformer_engine/common/fused_attn/utils.h
index d5cf450..25c295d 100644
--- a/transformer_engine/common/fused_attn/utils.h
+++ b/transformer_engine/common/fused_attn/utils.h
@@ -140,12 +140,12 @@ cudnn_frontend::DataType_t get_cudnn_fe_dtype(const transformer_engine::DType t)
 class cudnnExecutionPlanManager {
  public:
   static cudnnExecutionPlanManager &Instance() {
-    static thread_local cudnnExecutionPlanManager instance;
+    static cudnnExecutionPlanManager instance;
     return instance;
   }
 
   cudnnHandle_t GetCudnnHandle() {
-    static thread_local std::once_flag flag;
+    static std::once_flag flag;
     std::call_once(flag, [&] { cudnnCreate(&handle_); });
     return handle_;
   }
diff --git a/transformer_engine/jax/cpp_extensions/activation.py b/transformer_engine/jax/cpp_extensions/activation.py
index 47483c6..6e2100b 100644
--- a/transformer_engine/jax/cpp_extensions/activation.py
+++ b/transformer_engine/jax/cpp_extensions/activation.py
@@ -162,7 +162,7 @@ class ActLuPrimitive(BasePrimitive):
         act_lu infer_sharding_from_operands
         """
         del result_infos, act_enum  # Unused.
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-2], x_spec[-1]))
         return out_sharding
 
@@ -172,7 +172,7 @@ class ActLuPrimitive(BasePrimitive):
         act_lu partitioning
         """
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-2], x_spec[-1]))
 
@@ -300,7 +300,7 @@ class DActLuPrimitive(BasePrimitive):
         dact_lu infer_sharding_from_operands
         """
         del result_infos, act_enum  # Unused.
-        act_lu_out_spec = get_padded_spec(arg_infos[1])
+        act_lu_out_spec = get_padded_spec(arg_infos[1], mesh)
         dx_sharding = NamedSharding(mesh, PartitionSpec(*act_lu_out_spec))
         return dx_sharding
 
@@ -310,7 +310,7 @@ class DActLuPrimitive(BasePrimitive):
         dact_lu partition
         """
         del result_infos
-        dx_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        dx_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = dx_sharding
 
@@ -449,17 +449,17 @@ class ActLuFp8Primitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(out_dtype, act_enum, mesh, arg_infos, result_infos):
         del out_dtype, result_infos, act_enum
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-2], x_spec[-1]))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         return (out_sharding, amax_sharding)
 
     @staticmethod
     def partition(out_dtype, act_enum, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-2], x_spec[-1]))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = (out_sharding, amax_sharding)
 
diff --git a/transformer_engine/jax/cpp_extensions/attention.py b/transformer_engine/jax/cpp_extensions/attention.py
index d5b901c..29b5710 100644
--- a/transformer_engine/jax/cpp_extensions/attention.py
+++ b/transformer_engine/jax/cpp_extensions/attention.py
@@ -500,7 +500,7 @@ class FusedAttnFwdPrimitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(config, mesh, arg_infos, result_infos):
         del result_infos
-        q_spec = get_padded_spec(arg_infos[0])
+        q_spec = get_padded_spec(arg_infos[0], mesh)
         match config.qkv_layout:
             case NVTE_QKV_Layout.NVTE_BS3HD | NVTE_QKV_Layout.NVTE_T3HD:
                 # q_spec = (...batch, q_seqlen, head, hidden)
@@ -833,10 +833,10 @@ class FusedAttnBwdPrimitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(config, mesh, arg_infos, result_infos):
         del config, result_infos
-        q_spec = get_padded_spec(arg_infos[0])
-        k_spec = get_padded_spec(arg_infos[1])
-        v_spec = get_padded_spec(arg_infos[2])
-        bias_spec = get_padded_spec(arg_infos[3])
+        q_spec = get_padded_spec(arg_infos[0], mesh)
+        k_spec = get_padded_spec(arg_infos[1], mesh)
+        v_spec = get_padded_spec(arg_infos[2], mesh)
+        bias_spec = get_padded_spec(arg_infos[3], mesh)
         dq_sharding = NamedSharding(mesh, PartitionSpec(*q_spec))
         dk_sharding = NamedSharding(mesh, PartitionSpec(*k_spec))
         dv_sharding = NamedSharding(mesh, PartitionSpec(*v_spec))
@@ -846,10 +846,10 @@ class FusedAttnBwdPrimitive(BasePrimitive):
     @staticmethod
     def partition(config, mesh, arg_infos, result_infos):
         del result_infos
-        q_spec = get_padded_spec(arg_infos[0])
-        k_spec = get_padded_spec(arg_infos[1])
-        v_spec = get_padded_spec(arg_infos[2])
-        bias_spec = get_padded_spec(arg_infos[3])
+        q_spec = get_padded_spec(arg_infos[0], mesh)
+        k_spec = get_padded_spec(arg_infos[1], mesh)
+        v_spec = get_padded_spec(arg_infos[2], mesh)
+        bias_spec = get_padded_spec(arg_infos[3], mesh)
         dq_sharding = NamedSharding(mesh, PartitionSpec(*q_spec))
         dk_sharding = NamedSharding(mesh, PartitionSpec(*k_spec))
         dv_sharding = NamedSharding(mesh, PartitionSpec(*v_spec))
@@ -1144,10 +1144,10 @@ class FusedAttnCPWithAllGatherBwdPrimitive(FusedAttnBwdPrimitive):
         helper.check_supported()

         del result_infos
-        q_spec = get_padded_spec(arg_infos[0])
-        k_spec = get_padded_spec(arg_infos[1])
-        v_spec = get_padded_spec(arg_infos[2])
-        bias_spec = get_padded_spec(arg_infos[3])
+        q_spec = get_padded_spec(arg_infos[0], mesh)
+        k_spec = get_padded_spec(arg_infos[1], mesh)
+        v_spec = get_padded_spec(arg_infos[2], mesh)
+        bias_spec = get_padded_spec(arg_infos[3], mesh)
         dq_sharding = NamedSharding(mesh, PartitionSpec(*q_spec))
         dk_sharding = NamedSharding(mesh, PartitionSpec(*k_spec))
         dv_sharding = NamedSharding(mesh, PartitionSpec(*v_spec))
diff --git a/transformer_engine/jax/cpp_extensions/base.py b/transformer_engine/jax/cpp_extensions/base.py
index 3d88c1f..1fe4c15 100644
--- a/transformer_engine/jax/cpp_extensions/base.py
+++ b/transformer_engine/jax/cpp_extensions/base.py
@@ -104,6 +104,7 @@ def register_primitive(cls):
     inner_p.def_impl(partial(xla.apply_primitive, inner_p))
     inner_p.def_abstract_eval(cls.abstract)
     mlir.register_lowering(inner_p, cls.lowering, platform="cuda")
+    mlir.register_lowering(inner_p, cls.lowering, platform="legate")
     cls.inner_primitive = inner_p
 
     outer_p = core.Primitive(name_of_wrapper_p())
diff --git a/transformer_engine/jax/cpp_extensions/misc.py b/transformer_engine/jax/cpp_extensions/misc.py
index 58b8db4..e91b30a 100644
--- a/transformer_engine/jax/cpp_extensions/misc.py
+++ b/transformer_engine/jax/cpp_extensions/misc.py
@@ -18,7 +18,7 @@ from jax.interpreters.mlir import dtype_to_ir_type
 from transformer_engine.transformer_engine_jax import DType as TEDType
 from transformer_engine import transformer_engine_jax
 
-from ..sharding import get_padded_spec as te_get_padded_spec
+from ..sharding import te_get_padded_spec
 
 
 def te_dtype_to_jax_dtype(te_dtype):
@@ -81,14 +81,22 @@ def jax_dtype_to_te_dtype(jax_dtype):
     return converter.get(jax_dtype)
 
 
-def get_padded_spec(arg_info):
+def get_padded_spec(arg_info, mesh):
     """
     Get padded spec for partitioning from arguments' information
     """
     if arg_info.sharding is None:
         return te_get_padded_spec(None, arg_info.ndim)
-    ndim, spec = arg_info.ndim, arg_info.sharding.spec
-    return te_get_padded_spec(spec, ndim)
+
+    from jax._src.sharding_impls import PositionalSharding, parse_flatten_op_sharding
+    if isinstance(arg_info.sharding, PositionalSharding):
+        s = arg_info.sharding._to_xla_hlo_sharding(arg_info.ndim)
+        pspec = parse_flatten_op_sharding(s, mesh)[0]
+        spec = tuple(d for d in pspec)
+    else:
+        spec = arg_info.sharding.spec
+
+    return te_get_padded_spec(spec, arg_info.ndim)
 
 
 def check_valid_batch_dims(bdims):
diff --git a/transformer_engine/jax/cpp_extensions/normalization.py b/transformer_engine/jax/cpp_extensions/normalization.py
index e85f28a..88b315b 100644
--- a/transformer_engine/jax/cpp_extensions/normalization.py
+++ b/transformer_engine/jax/cpp_extensions/normalization.py
@@ -205,7 +205,7 @@ class LayerNormFwdPrimitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(zero_centered_gamma, epsilon, mesh, arg_infos, result_infos):
         del zero_centered_gamma, epsilon, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {LayerNormFwdPrimitive.name}! "
@@ -219,7 +219,7 @@ class LayerNormFwdPrimitive(BasePrimitive):
     @staticmethod
     def partition(zero_centered_gamma, epsilon, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec, g_spec, b_spec = map(get_padded_spec, arg_infos)
+        x_spec, g_spec, b_spec = map(partial(get_padded_spec, mesh=mesh), arg_infos)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {LayerNormFwdPrimitive.name}! "
@@ -485,14 +485,14 @@ class LayerNormBwdPrimitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(zero_centered_gamma, epsilon, mesh, arg_infos, result_infos):
         del zero_centered_gamma, epsilon, result_infos
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {LayerNormBwdPrimitive.name}! "
                 "Force to not shard the hidden dim, which might introduce extra collective ops, "
                 "and hurt performance."
             )
-        g_b_spec = get_padded_spec(arg_infos[4])
+        g_b_spec = get_padded_spec(arg_infos[4], mesh)
         if g_b_spec[-1] is not None:
             warnings.warn(
                 f"{LayerNormBwdPrimitive.name} does not support sharding of gradients "
@@ -507,14 +507,14 @@ class LayerNormBwdPrimitive(BasePrimitive):
     @staticmethod
     def partition(zero_centered_gamma, epsilon, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {LayerNormBwdPrimitive.name}! "
                 "Force to not shard the hidden dim, which might introduce extra collective ops, "
                 "and hurt performance."
             )
-        g_b_spec = get_padded_spec(arg_infos[4])
+        g_b_spec = get_padded_spec(arg_infos[4], mesh)
         if g_b_spec[-1] is not None:
             warnings.warn(
                 f"{LayerNormBwdPrimitive.name} does not support sharding of gradients "
@@ -533,8 +533,11 @@ class LayerNormBwdPrimitive(BasePrimitive):
             local_dx, local_dgamma, local_dbeta = LayerNormBwdPrimitive.impl(
                 dz, x, mu, rsigma, gamma, zero_centered_gamma=zero_centered_gamma, epsilon=epsilon
             )
-            global_dgamma = all_reduce_sum_along_dp_fsdp(local_dgamma, mesh)
-            global_dbeta = all_reduce_sum_along_dp_fsdp(local_dbeta, mesh)
+
+            #global_dgamma = all_reduce_sum_along_dp_fsdp(local_dgamma, mesh)
+            global_dgamma = local_dgamma
+            #global_dbeta = all_reduce_sum_along_dp_fsdp(local_dbeta, mesh)
+            global_dbeta = local_dbeta
             return local_dx, global_dgamma, global_dbeta
 
         return mesh, sharded_impl, out_shardings, arg_shardings
@@ -702,7 +705,7 @@ class RmsNormFwdPrimitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(epsilon, mesh, arg_infos, result_infos):
         del epsilon, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {RmsNormFwdPrimitive.name}! "
@@ -893,14 +896,14 @@ class RmsNormBwdPrimitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(epsilon, mesh, arg_infos, result_infos):
         del epsilon, result_infos
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {RmsNormBwdPrimitive.name}! "
                 "Force to not shard the hidden dim, which might introduce extra collective ops, "
                 "and hurt performance."
             )
-        g_spec = get_padded_spec(arg_infos[3])
+        g_spec = get_padded_spec(arg_infos[3], mesh)
         if g_spec[-1] is not None:
             warnings.warn(
                 f"{RmsNormBwdPrimitive.name} does not support sharding of parameter gamma "
@@ -913,14 +916,14 @@ class RmsNormBwdPrimitive(BasePrimitive):
     @staticmethod
     def partition(epsilon, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {RmsNormBwdPrimitive.name}! "
                 "Force to not shard the hidden dim, which might introduce extra collective ops, "
                 "and hurt performance."
             )
-        g_spec = get_padded_spec(arg_infos[3])
+        g_spec = get_padded_spec(arg_infos[3], mesh)
         if g_spec[-1] is not None:
             warnings.warn(
                 f"{RmsNormBwdPrimitive.name} does not support sharding of parameter gamma "
@@ -1169,7 +1172,7 @@ class LayerNormFwdFp8Primitive(BasePrimitive):
         out_dtype, zero_centered_gamma, epsilon, mesh, arg_infos, result_infos
     ):
         del out_dtype, zero_centered_gamma, epsilon, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {LayerNormFwdPrimitive.name}! "
@@ -1179,15 +1182,15 @@ class LayerNormFwdFp8Primitive(BasePrimitive):
 
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-1], None))
         mu_sharding = rsigma_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-1]))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[3])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[3], mesh)))
         return (out_sharding, mu_sharding, rsigma_sharding, amax_sharding)
 
     @staticmethod
     def partition(out_dtype, zero_centered_gamma, epsilon, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
-        g_spec = get_padded_spec(arg_infos[1])
-        b_spec = get_padded_spec(arg_infos[2])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
+        g_spec = get_padded_spec(arg_infos[1], mesh)
+        b_spec = get_padded_spec(arg_infos[2], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {LayerNormFwdFp8Primitive.name}! "
@@ -1209,9 +1212,9 @@ class LayerNormFwdFp8Primitive(BasePrimitive):
         b_sharding = NamedSharding(mesh, PartitionSpec(None))
         out_sharding = x_sharding
         mu_sharding = rsigma_sharding = NamedSharding(
-            mesh, PartitionSpec(*get_padded_spec(arg_infos[0])[:-1])
+            mesh, PartitionSpec(*get_padded_spec(arg_infos[0], mesh)[:-1])
         )
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[3])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[3], mesh)))
         fp8_meta_sharding = amax_sharding
         arg_shardings = (x_sharding, g_sharding, b_sharding) + (fp8_meta_sharding,) * 3
         out_shardings = (out_sharding, mu_sharding, rsigma_sharding, amax_sharding)
@@ -1440,7 +1443,7 @@ class RmsNormFwdFp8Primitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(out_dtype, epsilon, mesh, arg_infos, result_infos):
         del out_dtype, epsilon, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {RmsNormFwdFp8Primitive.name}! "
@@ -1449,14 +1452,14 @@ class RmsNormFwdFp8Primitive(BasePrimitive):
             )
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-1], None))
         rsigma_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-1]))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2], mesh)))
         return (out_sharding, rsigma_sharding, amax_sharding)
 
     @staticmethod
     def partition(out_dtype, epsilon, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
-        g_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
+        g_spec = get_padded_spec(arg_infos[1], mesh)
         if x_spec[-1] is not None:
             warnings.warn(
                 f"Does not support to shard hidden dim in {RmsNormFwdFp8Primitive.name}! "
@@ -1471,8 +1474,8 @@ class RmsNormFwdFp8Primitive(BasePrimitive):
         x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec[:-1], None))
         g_sharding = NamedSharding(mesh, PartitionSpec(None))
         out_sharding = x_sharding
-        rsigma_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[0])[:-1]))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2])))
+        rsigma_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[0], mesh)[:-1]))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2], mesh)))
         fp8_meta_sharding = amax_sharding
         arg_shardings = (x_sharding, g_sharding) + (fp8_meta_sharding,) * 3
         out_shardings = (out_sharding, rsigma_sharding, amax_sharding)
diff --git a/transformer_engine/jax/cpp_extensions/quantization.py b/transformer_engine/jax/cpp_extensions/quantization.py
index 48bf4d9..4dfe50a 100644
--- a/transformer_engine/jax/cpp_extensions/quantization.py
+++ b/transformer_engine/jax/cpp_extensions/quantization.py
@@ -139,17 +139,17 @@ class CastFP8Primitive(BasePrimitive):
     @staticmethod
     def infer_sharding_from_operands(out_dtype, mesh, arg_infos, result_infos):
         del out_dtype, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         return (casted_x_sharding, amax_sharding)
 
     @staticmethod
     def partition(out_dtype, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = (casted_x_sharding, amax_sharding)
 
diff --git a/transformer_engine/jax/cpp_extensions/softmax.py b/transformer_engine/jax/cpp_extensions/softmax.py
index bf92c00..2e07614 100644
--- a/transformer_engine/jax/cpp_extensions/softmax.py
+++ b/transformer_engine/jax/cpp_extensions/softmax.py
@@ -189,7 +189,7 @@ class SoftmaxPrimitive(BasePrimitive):
         softmax_forward infer_sharding_from_operands
         """
         del scale_factor, result_infos  # Unused.
-        logits_spec = get_padded_spec(arg_infos[0])
+        logits_spec = get_padded_spec(arg_infos[0], mesh)
         if logits_spec[-1] is not None:
             warnings.warn(
                 f"Sharding the hidden dimension is not supported in {cls.name}! "
@@ -205,7 +205,7 @@ class SoftmaxPrimitive(BasePrimitive):
         softmax_forward partitioning
         """
         del result_infos
-        logits_spec = get_padded_spec(arg_infos[0])
+        logits_spec = get_padded_spec(arg_infos[0], mesh)
         if logits_spec[-1] is not None:
             warnings.warn(
                 f"Sharding the hidden dimension is not supported in {cls.name}! "
@@ -301,7 +301,7 @@ class SoftmaxPrimitive(BasePrimitive):
         softmax_backward infer_sharding_from_operands
         """
         del scale_factor, result_infos  # Unused.
-        dz_spec = get_padded_spec(arg_infos[0])
+        dz_spec = get_padded_spec(arg_infos[0], mesh)
         if dz_spec[-1] is not None:
             warnings.warn(
                 f"Sharding the hidden dimension is not supported in {cls.name}! "
@@ -318,8 +318,8 @@ class SoftmaxPrimitive(BasePrimitive):
         """
         del result_infos
 
-        dz_spec = get_padded_spec(arg_infos[0])
-        softmax_out_spec = get_padded_spec(arg_infos[1])
+        dz_spec = get_padded_spec(arg_infos[0], mesh)
+        softmax_out_spec = get_padded_spec(arg_infos[1], mesh)
         if dz_spec[-1] is not None or softmax_out_spec[-1] is not None:
             warnings.warn(
                 f"Sharding the hidden dimension is not supported in {cls.name}! "
diff --git a/transformer_engine/jax/cpp_extensions/transpose.py b/transformer_engine/jax/cpp_extensions/transpose.py
index 963d7f0..c304580 100644
--- a/transformer_engine/jax/cpp_extensions/transpose.py
+++ b/transformer_engine/jax/cpp_extensions/transpose.py
@@ -170,7 +170,7 @@ class TransposePrimitive(BasePrimitive):
         static_axis_boundary, transpose_axis_boundary, mesh, arg_infos, result_infos
     ):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
         return transposed_x_sharding
@@ -178,7 +178,7 @@ class TransposePrimitive(BasePrimitive):
     @staticmethod
     def partition(static_axis_boundary, transpose_axis_boundary, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
@@ -360,11 +360,11 @@ class CastTransposePrimitive(BasePrimitive):
         out_dtype, static_axis_boundary, transpose_axis_boundary, mesh, arg_infos, result_infos
     ):
         del out_dtype, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         casted_transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         return (casted_x_sharding, casted_transposed_x_sharding, amax_sharding)
 
     @staticmethod
@@ -372,11 +372,11 @@ class CastTransposePrimitive(BasePrimitive):
         out_dtype, static_axis_boundary, transpose_axis_boundary, mesh, arg_infos, result_infos
     ):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         casted_transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = (casted_x_sharding, casted_transposed_x_sharding, amax_sharding)
 
@@ -603,14 +603,14 @@ class DBiasCastTransposePrimitive(BasePrimitive):
         out_dtype, static_axis_boundary, transpose_axis_boundary, mesh, arg_infos, result_infos
     ):
         del out_dtype, result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         tranposed_out_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
         dbias_shaprding = NamedSharding(
             mesh, PartitionSpec(*x_spec[: static_axis_boundary + 1], x_spec[-1])
         )
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         return (out_sharding, tranposed_out_sharding, dbias_shaprding, amax_sharding)
 
     @staticmethod
@@ -618,7 +618,7 @@ class DBiasCastTransposePrimitive(BasePrimitive):
         out_dtype, static_axis_boundary, transpose_axis_boundary, mesh, arg_infos, result_infos
     ):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[0])
+        x_spec = get_padded_spec(arg_infos[0], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         casted_transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
@@ -627,7 +627,7 @@ class DBiasCastTransposePrimitive(BasePrimitive):
             mesh, PartitionSpec(*x_spec[: static_axis_boundary + 1], x_spec[-1])
         )
 
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = (
             casted_x_sharding,
@@ -928,14 +928,14 @@ class DActLuDBiasCastTransposePrimitive(BasePrimitive):
         result_infos,
     ):
         del out_dtype, result_infos, act_enum
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         tranposed_out_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
         dbias_shaprding = NamedSharding(
             mesh, PartitionSpec(*x_spec[: static_axis_boundary + 1], x_spec[-1])
         )
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2], mesh)))
         return (out_sharding, tranposed_out_sharding, dbias_shaprding, amax_sharding)
 
     @staticmethod
@@ -949,7 +949,7 @@ class DActLuDBiasCastTransposePrimitive(BasePrimitive):
         result_infos,
     ):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, transpose_axis_boundary)
         casted_transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
@@ -958,7 +958,7 @@ class DActLuDBiasCastTransposePrimitive(BasePrimitive):
             mesh, PartitionSpec(*x_spec[: static_axis_boundary + 1], x_spec[-1])
         )
 
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = (
             casted_x_sharding,
@@ -1195,22 +1195,22 @@ class DgatedActLuCastTransposePrimitive(BasePrimitive):
         out_dtype, static_axis_boundary, act_enum, mesh, arg_infos, result_infos
     ):
         del out_dtype, result_infos, act_enum
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         out_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, -2)
         tranposed_out_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2], mesh)))
         return (out_sharding, tranposed_out_sharding, amax_sharding)
 
     @staticmethod
     def partition(out_dtype, static_axis_boundary, act_enum, mesh, arg_infos, result_infos):
         del result_infos
-        x_spec = get_padded_spec(arg_infos[1])
+        x_spec = get_padded_spec(arg_infos[1], mesh)
         casted_x_sharding = NamedSharding(mesh, PartitionSpec(*x_spec))
         xt_spec = multidim_transpose(x_spec, static_axis_boundary, -2)
         casted_transposed_x_sharding = NamedSharding(mesh, PartitionSpec(*xt_spec))
 
-        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2])))
+        amax_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[2], mesh)))
         arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
         out_shardings = (casted_x_sharding, casted_transposed_x_sharding, amax_sharding)
 
diff --git a/transformer_engine/jax/csrc/utils.h b/transformer_engine/jax/csrc/utils.h
index 32de33b..12b3cd3 100644
--- a/transformer_engine/jax/csrc/utils.h
+++ b/transformer_engine/jax/csrc/utils.h
@@ -34,7 +34,7 @@ uint32_t GetRuntimeNumSegments(void *cu_seqlen, void *workspace, size_t len, cud
 class cudaDevicePropertiesManager {
  public:
   static cudaDevicePropertiesManager &Instance() {
-    static thread_local cudaDevicePropertiesManager instance;
+    static cudaDevicePropertiesManager instance;
     return instance;
   }
 
@@ -66,7 +66,7 @@ class cudaDevicePropertiesManager {
 class FusedAttnOffsetManager {
  public:
   static FusedAttnOffsetManager &Instance() {
-    static thread_local FusedAttnOffsetManager instance;
+    static FusedAttnOffsetManager instance;
     return instance;
   }
 
diff --git a/transformer_engine/jax/fp8.py b/transformer_engine/jax/fp8.py
index 5df8ce4..53b960e 100644
--- a/transformer_engine/jax/fp8.py
+++ b/transformer_engine/jax/fp8.py
@@ -78,7 +78,7 @@ def _format2dtypes(format_: Format):
 # would sum gradients from all micro-batches, and this is not the expected
 # behavior for FP8 meta. Instead, the summation of FP8 meta gradients should
 # be "MAX".
-FlaxFloatMeta32 = fp8_ops.fm32
+# FlaxFloatMeta32 = fp8_ops.fm32
 
 
 class FP8MetaPackage:
diff --git a/transformer_engine/jax/praxis/transformer.py b/transformer_engine/jax/praxis/transformer.py
index 2651144..c437f6a 100644
--- a/transformer_engine/jax/praxis/transformer.py
+++ b/transformer_engine/jax/praxis/transformer.py
@@ -151,6 +151,7 @@ class MultiHeadAttention(TransformerEngineBaseLayer):
     scale_attn_logits: bool = False
     scaled_query_init: bool = True
     float32_logits: bool = False
+    atten_logit_cap: float = 0.
 
     # Deprecated parameters
     num_heads: Optional[int] = None
diff --git a/transformer_engine/jax/sharding.py b/transformer_engine/jax/sharding.py
index a14a838..16d87ab 100644
--- a/transformer_engine/jax/sharding.py
+++ b/transformer_engine/jax/sharding.py
@@ -123,7 +123,7 @@ def get_all_mesh_axes():
     return mesh.axis_names
 
 
-def get_padded_spec(spec, ndim):
+def te_get_padded_spec(spec, ndim):
     """
     Get padded spec for partitioning from arguments' information
     """
@@ -142,6 +142,7 @@ def lax_paral_op(
     if mesh_resource is not None:
         _, resource = _get_mesh_info(mesh_resource, mesh)
         return ops(x, resource, **kwargs)
+
     return x
 
 
diff --git a/transformer_engine/paddle/csrc/common.h b/transformer_engine/paddle/csrc/common.h
index 6ce2504..60a0cad 100644
--- a/transformer_engine/paddle/csrc/common.h
+++ b/transformer_engine/paddle/csrc/common.h
@@ -142,7 +142,7 @@ inline NVTE_Fused_Attn_Backend get_fused_attn_backend(
 class cudaDevicePropertiesManager {
  public:
   static cudaDevicePropertiesManager &Instance() {
-    static thread_local cudaDevicePropertiesManager instance;
+    static cudaDevicePropertiesManager instance;
     return instance;
   }
 
