diff --git a/MaxText/configs/base.yml b/MaxText/configs/base.yml
index 81ec1d54..8e3064a7 100644
--- a/MaxText/configs/base.yml
+++ b/MaxText/configs/base.yml
@@ -284,11 +284,11 @@ jax_cache_dir: "~/jax_cache"
 hardware: 'tpu' # Supported hardware types are 'tpu', 'gpu', 'gpu_multiprocess' and 'cpu'
 
 # Parallelism
-mesh_axes: ['data', 'stage', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']
+mesh_axes: ['stage', 'data', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']
 logical_axis_rules: [
                       ['activation_batch', ['data', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_batch_no_exp', ['data', 'fsdp', 'fsdp_transpose']],
-                      ['activation_embed_and_logits_batch', ['data', 'stage', 'fsdp', 'fsdp_transpose', 'expert']],
+                      ['activation_embed_and_logits_batch', ['stage', 'data', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence','autoregressive']],
                       ['activation_kv_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
                       ['activation_length', ['sequence', 'context']],
@@ -349,7 +349,7 @@ logical_axis_rules: [
                       ['paged_kv_head_dim_size', []],
                     ]
 # Axes used for DCN must be earlier in this list than ICI, see (b/339009148) for details
-data_sharding: [['data', 'stage', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']]
+data_sharding: [['fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']]
 input_data_sharding_logical_axes: ['activation_embed_and_logits_batch', 'activation_norm_length']
 
 # sharding tolerance: float between 0.0 and 1.0 representing the allowed percentage of non-sharded parameters.
@@ -514,7 +514,7 @@ data_shuffle_seed: 0
 init_weights_seed: 0
 
 # You may disable clipping by setting gradient_clipping_threshold to zero.
-gradient_clipping_threshold: 1.0
+gradient_clipping_threshold: 0.0
 
 # Instead of updating the weights every step, you may effectively use a larger
 # batch by accumulating the gradient over a set of steps.
diff --git a/MaxText/configs/inference.yml b/MaxText/configs/inference.yml
index 3b830949..c3302296 100644
--- a/MaxText/configs/inference.yml
+++ b/MaxText/configs/inference.yml
@@ -3,7 +3,7 @@ base_config: "base.yml"
 logical_axis_rules: [
                       ['activation_batch', ['data', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_batch_no_exp', ['data', 'fsdp', 'fsdp_transpose']],
-                      ['activation_embed_and_logits_batch', ['data', 'stage', 'fsdp', 'fsdp_transpose', 'expert']],
+                      ['activation_embed_and_logits_batch', ['stage', 'data', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
                       ['activation_kv_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
                       ['activation_length', ['context_autoregressive', 'sequence']],
diff --git a/MaxText/configs/models/gpt3-24layers.yml b/MaxText/configs/models/gpt3-24layers.yml
new file mode 100644
index 00000000..5b4dafbb
--- /dev/null
+++ b/MaxText/configs/models/gpt3-24layers.yml
@@ -0,0 +1,43 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing software
+# distributed under the License is distributed on an "AS IS" BASIS
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# model config for gpt3-175b
+
+base_emb_dim: 12288
+base_num_query_heads: 96
+base_num_kv_heads: 96
+base_mlp_dim: 49152
+base_num_decoder_layers: 24
+head_dim: 128
+trainable_position_size: 16384
+mlp_activations: ["gelu"]
+vocab_size: 50304
+enable_dropout: False
+logits_via_embedding: True
+normalize_embedding_logits: False
+logits_dot_in_fp32: False
+normalization_layer_epsilon: 1.e-05
+use_iota_embed: True
+fused_qkv: True
+opt_type: "adam_pax"
+decoder_block: "gpt3"
+dataset_type: "synthetic"
+gradient_clipping_threshold: 0.
+adam_b1: 0.9
+adam_b2: 0.95
+adam_eps: 1.e-8
+adam_weight_decay: 0.1
+checkpoint_period: 10_000
+target_eval_loss: 2.69
+eval_per_device_batch_size: 1.0
diff --git a/MaxText/configs/models/gpt3-xl.yml b/MaxText/configs/models/gpt3-xl.yml
new file mode 100644
index 00000000..fa0946c1
--- /dev/null
+++ b/MaxText/configs/models/gpt3-xl.yml
@@ -0,0 +1,55 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing software
+# distributed under the License is distributed on an "AS IS" BASIS
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# model config for gpt3-6b
+
+base_emb_dim: 2048
+base_num_query_heads: 24
+base_num_kv_heads: 24
+base_mlp_dim: 2048
+base_num_decoder_layers: 24
+head_dim: 128
+max_target_length: 1024
+trainable_position_size: 16384
+mlp_activations: ["gelu"]
+vocab_size: 32768
+enable_dropout: False
+logits_via_embedding: True
+normalize_embedding_logits: False
+logits_dot_in_fp32: False
+normalization_layer_epsilon: 1.e-05
+use_iota_embed: True
+fused_qkv: True
+opt_type: "adam_pax"
+decoder_block: "gpt3"
+#dataset_path: "gs://mlperf-llm-public2"
+#dataset_name: "c4/en:3.0.4"
+#eval_dataset_name: "c4/en:3.0.5"
+dataset_type: hf
+hf_path: '/home/jwilke/Programs/workspaces/realm-jax-refactor/workflows-update/maxtext/wikitext'
+hf_data_dir: 'wikitext-103-v1'
+hf_train_files: '/home/jwilke/Programs/workspaces/realm-jax-refactor/workflows-update/maxtext/wikitext/wikitext-103-v1/train*.parquet'
+# set eval_interval > 0 to use the specified eval dataset, otherwise, only metrics on the train set will be calculated.
+eval_interval: 10000
+hf_eval_split: 'validation'
+hf_eval_files: ''
+# for HF pipeline, tokenizer_path can be a path in HuggingFace Hub,
+# or a local path containing tokenizer in a format supported by transformers.AutoTokenizer
+tokenizer_path: 'google-t5/t5-large'  # for using https://huggingface.co/google-t5/t5-large
+hf_access_token: ''  # provide token if using gated dataset or tokenizer
+gradient_clipping_threshold: 1.
+adam_b1: 0.9
+adam_b2: 0.95
+adam_eps: 1.e-8
+adam_weight_decay: 0.1
diff --git a/MaxText/configs/models/mixtral-small.yml b/MaxText/configs/models/mixtral-small.yml
new file mode 100644
index 00000000..3eaf08f1
--- /dev/null
+++ b/MaxText/configs/models/mixtral-small.yml
@@ -0,0 +1,32 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# model config for mixtral-small
+# tokenizer_path is assets/tokenizer.mistral-v1
+
+base_emb_dim: 128
+base_num_query_heads: 16
+base_num_kv_heads: 8
+base_mlp_dim: 512
+base_num_decoder_layers: 8
+head_dim: 32
+mlp_activations: ["silu","linear"]
+vocab_size: 256
+enable_dropout: False
+logits_via_embedding: False
+normalization_layer_epsilon: 1.0e-5
+num_experts: 4
+num_experts_per_tok: 2
+rope_max_timescale: 1_000_000
+decoder_block: "mixtral"
\ No newline at end of file
diff --git a/MaxText/input_pipeline/_hf_data_processing.py b/MaxText/input_pipeline/_hf_data_processing.py
index 3fabdc79..2bd24d3c 100644
--- a/MaxText/input_pipeline/_hf_data_processing.py
+++ b/MaxText/input_pipeline/_hf_data_processing.py
@@ -187,7 +187,7 @@ def preprocessing_pipeline(
       data_source=dataset,
       operations=operations,
       sampler=dummy_index_sampler,
-      worker_count=grain_worker_count,  # only supports <=1 for now, more workers results in duplicated data
+      worker_count=0,  # only supports one worker for now, more workers results in duplicated data
       worker_buffer_size=1,
       read_options=grain.ReadOptions(num_threads=num_threads, prefetch_buffer_size=128),
   )
diff --git a/MaxText/input_pipeline/input_pipeline_interface.py b/MaxText/input_pipeline/input_pipeline_interface.py
index d4328ff9..653d889f 100644
--- a/MaxText/input_pipeline/input_pipeline_interface.py
+++ b/MaxText/input_pipeline/input_pipeline_interface.py
@@ -158,7 +158,7 @@ def make_mixed_iterator(
 def create_data_iterator(config: pyconfig.HyperParameters, mesh):
   """create data iterator"""
   if config.dataset_type == "synthetic":
-    return SyntheticDataIterator(config, mesh), None
+    return SyntheticDataIterator(config, mesh), SyntheticDataIterator(config, mesh)
 
   process_indices_train = get_process_loading_real_data(
       config.data_sharding,
diff --git a/MaxText/layers/embeddings.py b/MaxText/layers/embeddings.py
index 9a7b4b0e..2e4e5680 100644
--- a/MaxText/layers/embeddings.py
+++ b/MaxText/layers/embeddings.py
@@ -100,6 +100,14 @@ class Embed(nn.Module):
       output = jnp.dot(one_hot, jnp.asarray(self.embedding, self.dtype))
     else:
       output = jnp.asarray(self.embedding, self.dtype)[inputs]
+
+      # encourage XLA compiler to replicate each batch groups embeddings
+      # prior to sharding on remaining dimensions for faster SPMD code
+      output = nn.with_logical_constraint(
+        output, ("activation_embed_and_logits_batch", None, None)
+      )
+
+
     output = nn.with_logical_constraint(
         output, ("activation_embed_and_logits_batch", "activation_length", "activation_embed")
     )
diff --git a/MaxText/layers/models.py b/MaxText/layers/models.py
index f542c581..0127bf37 100644
--- a/MaxText/layers/models.py
+++ b/MaxText/layers/models.py
@@ -201,7 +201,7 @@ class Decoder(nn.Module):
     """Initialize decoder layer."""
     self.decoder_layer = self.get_decoder_layers()
     self.norm_layer = self.get_norm_layer()
-    if self.config.using_pipeline_parallelism:
+    if False and self.config.using_pipeline_parallelism:
       pipeline_stage_module = self.get_pipeline_stage_module(self.decoder_layer)
       remat_policy = self.get_remat_policy()
       self.pipeline_module = pipeline.Pipeline(
@@ -498,7 +498,7 @@ class Decoder(nn.Module):
     policy = self.get_remat_policy()
     RemattedBlockLayers = self.set_remat_policy(self.decoder_layer, policy)
 
-    if cfg.using_pipeline_parallelism:
+    if False and cfg.using_pipeline_parallelism:
       if cfg.pipeline_fsdp_ag_once:
         partition_spec = self.pipeline_module.get_weight_sharding(
             y, decoder_segment_ids, decoder_positions, deterministic, model_mode
diff --git a/MaxText/layers/moe.py b/MaxText/layers/moe.py
index 2590b011..a491b4e2 100644
--- a/MaxText/layers/moe.py
+++ b/MaxText/layers/moe.py
@@ -24,7 +24,6 @@ import numpy as np
 
 from jax import lax
 from jax.ad_checkpoint import checkpoint_name
-from jax.experimental import shard_map
 from jax.sharding import Mesh
 import jax
 import jax.numpy as jnp
@@ -45,6 +44,8 @@ from MaxText.layers.attentions import NdInitializer, nd_dense_init
 from MaxText.layers.initializers import default_bias_init
 from MaxText.layers.quantizations import AqtQuantization as Quant
 
+from multimesh.jax import shard_map
+
 
 DISPATCH = "dispatch"
 COMBINE = "combine"
@@ -607,7 +608,7 @@ class RoutedMoE(nn.Module):
       wo_pspec = aqt_tensor.partition_spec(wo_pspec, (1,), wo_kernel.dtype, use_bias=False)
 
     @functools.partial(
-        shard_map.shard_map,
+        shard_map,
         mesh=self.mesh,
         in_specs=(input_partition_pspec, gate_logits_pspec, pre_bias_logits_pspec, w0_pspec, w1_pspec, wo_pspec),
         out_specs=(nn.logical_to_mesh_axes((batch_logical_axis, None, "activation_embed"))),
@@ -636,9 +637,9 @@ class RoutedMoE(nn.Module):
           # Note: Reducing buffer increase the risk of token dropping under unbalanced distribution.
           buffer_size = int(
               num_expert_parallelism
-              * self.config.per_device_batch_size
+              * batch_size
               * self.config.max_target_length
-              * self.config.num_experts_per_tok
+              * min(self.config.num_experts_per_tok, local_expert_size)
           )
           output_shape = jnp.zeros((buffer_size, self.config.emb_dim), dtype=x.dtype)
 
@@ -723,7 +724,6 @@ class RoutedMoE(nn.Module):
       )
 
       return output, None
-
     return wrapper(inputs, gate_logits, pre_bias_logits, w0_kernel, w1_kernel, wo_kernel)
 
   def reshape_and_update_weights(self, weights, indices):
diff --git a/MaxText/max_utils.py b/MaxText/max_utils.py
index 967994ee..64ffa4ad 100644
--- a/MaxText/max_utils.py
+++ b/MaxText/max_utils.py
@@ -32,7 +32,7 @@ import numpy as np
 import jax
 import jax.numpy as jnp
 from jax.experimental import mesh_utils
-
+from jax.experimental.pjit import pjit
 import flax
 
 import psutil
@@ -116,8 +116,9 @@ def close_summary_writer(summary_writer):
 
 
 def add_text_to_summary_writer(key, value, summary_writer):
-  """Writes given key-value pair to tensorboard as text/summary."""
-  if jax.process_index() == 0:
+  """Writes given key-value pair to tensorboard as text/summary"""
+  #if jax.process_index() == 0:
+  if False:
     summary_writer.add_text(key, value)
 
 
@@ -162,6 +163,12 @@ def maybe_initialize_jax_distributed_system(raw_keys):
       initialize_jax_for_tpu_with_emergency_checkpointing(raw_keys)
     max_logging.log("Jax distributed system initialized!")
 
+  max_logging.log(f"JAX process: {jax.process_index()} / {jax.process_count()}")
+  max_logging.log(f"JAX devices: {jax.devices()}")
+  max_logging.log(f"JAX local devices: {jax.local_devices()}")
+  max_logging.log(f"jax.device_count(): {jax.device_count()}")
+  max_logging.log(f"jax.local_device_count(): {jax.local_device_count()}")
+  max_logging.log(f"jax.process_count(): {jax.process_count()}")
 
 def initialize_jax_for_gpu(raw_keys):
   """Jax distributed initialize for GPUs."""
@@ -182,18 +189,18 @@ def initialize_jax_for_cpu(raw_keys):
   coordinator_ip_address = get_coordinator_ip_address()
   coordinator_address = coordinator_ip_address + ":1234"  # JAX coordinator port used in XPK
   # Env variables to be set in XPK or otherwise
-  job_index = int(os.environ.get("JOB_INDEX"))
-  job_completion_index = int(os.environ.get("JOB_COMPLETION_INDEX"))
-  processes_in_job = int(os.environ.get("PROCESSES_IN_JOB"))
-  pid = job_index * processes_in_job + job_completion_index
-  max_logging.log(f" Jax process id is {pid} ")
-  # Explicit initialize is needed only for CPUs
-  jax.distributed.initialize(
-      coordinator_address=coordinator_address,
-      process_id=pid,
-      num_processes=int(os.environ.get("JAX_PROCESS_COUNT")),
-      initialization_timeout=raw_keys["jax_distributed_initialization_timeout"],
-  )
+  if (job_index := os.environ.get("JOB_INDEX")) is not None:
+    job_completion_index = int(os.environ.get("JOB_COMPLETION_INDEX"))
+    processes_in_job = int(os.environ.get("PROCESSES_IN_JOB"))
+    pid = job_index * processes_in_job + job_completion_index
+    max_logging.log(f" Jax process id is {pid} ")
+    # Explicit initialize is needed only for CPUs
+    jax.distributed.initialize(
+        coordinator_address=coordinator_address,
+        process_id=pid,
+        num_processes=int(os.environ.get("JAX_PROCESS_COUNT")),
+    )
+
 
 
 def _wait_for_file_to_disappear(f, timeout=300):
diff --git a/MaxText/maxtext_utils.py b/MaxText/maxtext_utils.py
index d6c997ee..e84e61f7 100644
--- a/MaxText/maxtext_utils.py
+++ b/MaxText/maxtext_utils.py
@@ -28,11 +28,18 @@ from flax.training import train_state
 import numpy as np
 
 from jax.experimental import mesh_utils
+from jax.experimental.pjit import pjit, AUTO
 from jax.experimental.serialize_executable import deserialize_and_load
+
 from jax.sharding import PartitionSpec as P
 import jax
 import jax.numpy as jnp
 
+try:
+  from multimesh.jax import MultiMesh
+except ImportError:
+  pass
+
 import optax
 
 import orbax.checkpoint.experimental.emergency.checkpoint_manager as emergency_checkpoint_manager
@@ -50,6 +57,105 @@ OVERWRITE_WITH_GRADIENT = "_overwrite_with_gradient"
 NUM_IMAGES_PER_SEQUENCE = 1
 NUM_IMAGE_CHANNELS = 3
 
+def bind_mesh(pjitted_fn, global_mesh: jax.sharding.Mesh):
+  """Wraps a pjitted_fn with a mesh context."""
+
+  def call(*args):
+    with global_mesh:
+      return pjitted_fn(*args)
+
+  def lower(*args, **kwargs):
+    with global_mesh:
+      return pjitted_fn.lower(*args, **kwargs)
+
+  call.lower = lower
+  return call
+
+
+
+def jit_and_compile(
+    func,
+    func_input_args,
+    func_input_kwargs,
+    mesh,
+    in_shardings,
+    out_shardings,
+    static_argnums,
+    donate_argnums,
+):
+  """Jit, lower, and compile func."""
+  jitted = pjit(
+      func,
+      in_shardings=in_shardings,
+      out_shardings=out_shardings,
+      static_argnums=static_argnums,
+      donate_argnums=donate_argnums,
+  )
+
+  jitted = bind_mesh(jitted, mesh)
+  lowered = jitted.lower(*func_input_args, **func_input_kwargs)
+  compiled = lowered.compile()
+
+  sharding_str_arr = []
+  def check(inp, inp_sharding, out):
+    sharding_str_arr.append(f"shape={inp.shape} input={inp_sharding} output={out}")
+    if not inp_sharding.is_equivalent_to(out, len(inp.shape)):
+      raise Exception("\n".join(sharding_str_arr) + " mismatch")
+
+  jax.tree_util.tree_map(check, func_input_args[0], compiled.input_shardings[0][0], compiled.output_shardings[0])
+
+  return compiled
+
+def get_and_compile_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config, shaped_train_args):
+  """Get the shardings (both state and data) for train_step"""
+  mesh = MultiMesh(global_mesh=mesh)
+
+  def canonicalize_sharding(p):
+    if isinstance(p, jax.sharding.GSPMDSharding):
+      return p
+    return AUTO(mesh)
+
+  state_mesh_shardings = jax.tree_util.tree_map(canonicalize_sharding, state_mesh_annotations)
+
+  functional_train = get_functional_train_step(train_step, model, config, state_mesh_shardings)
+  functional_train.__name__ = "train_step"
+  data_pspec = P(*config.data_sharding)
+
+  data_sharding = jax.tree_util.tree_map(lambda p: jax.sharding.NamedSharding(mesh, p), data_pspec)
+  in_shardings = (state_mesh_shardings, data_sharding, None)  # State, batch, rng
+
+  out_shardings = (state_mesh_shardings, None)
+
+   # get shapes from config
+  #shaped_train_args, shaped_train_kwargs = get_shaped_inputs(mesh, model, config)
+  shaped_train_kwargs = {}
+
+  static_argnums = ()  # We partial out the static argnums of model and config
+  donate_argnums = 0  # This is the index of the state - we allow the compiler to make use of this memory.
+
+  from jax.lax import with_sharding_constraint
+
+  def autoshard_train_fn(train_state, *args):
+    sharded_train_state = jax.tree_util.tree_map(lambda x, y: with_sharding_constraint(x, y), train_state, state_mesh_annotations)
+    return functional_train(sharded_train_state, *args)
+
+  with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
+    compiled = jit_and_compile(
+        autoshard_train_fn,
+        shaped_train_args,
+        shaped_train_kwargs,
+        mesh,
+        in_shardings,
+        out_shardings,
+        static_argnums,
+        donate_argnums,
+    )
+
+  # extract in/out-sharding from compiled function
+  in_shardings = compiled.input_shardings[0]
+  out_shardings = compiled.output_shardings[0]
+
+  return compiled, in_shardings, out_shardings, static_argnums, donate_argnums
 
 def get_input_data_sharding(config, mesh):
   """Get the input data sharding for the model"""
@@ -76,7 +182,14 @@ def get_functional_eval_with_signature(eval_step, mesh, state_mesh_shardings, mo
   """Get the shardings (both state and data) for eval_step"""
   functional_eval = get_functional_eval_step(eval_step, model, config)
   functional_eval.__name__ = "eval_step"
-  data_sharding = get_input_data_sharding(config, mesh)
+
+  # data_sharding = get_input_data_sharding(config, mesh)
+
+  # MM-Jax: reverting to data_sharding because that is the sharding that 
+  # all data loaders generate, and unlike vanilla maxtext, we are not
+  # hitting a reshard from input_data_sharding_logical_axes to data_sharding
+  data_sharding = jax.sharding.NamedSharding(mesh, P(*config.data_sharding))
+
   in_shardings = (state_mesh_shardings, data_sharding, None)  # State, batch, rng
   out_shardings = None  # metrics
   static_argnums = ()  # We partial out the static argnums of model, config
@@ -508,6 +621,50 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
   state = max_utils.unbox_logicallypartioned(state)
   return state, state_mesh_annotations
 
+def bind_mesh(pjitted_fn, global_mesh: jax.sharding.Mesh):
+  """Wraps a pjitted_fn with a mesh context."""
+
+  def call(*args):
+    with global_mesh:
+      return pjitted_fn(*args)
+
+  def lower(*args, **kwargs):
+    with global_mesh:
+      return pjitted_fn.lower(*args, **kwargs)
+
+  call.lower = lower
+  return call
+
+
+def initialize_state_with_sharding(init_state_partial, mesh, state_sharding, rng_sharding, rng):
+
+  with mesh:
+    jitted = pjit(
+        init_state_partial,
+        in_shardings=rng_sharding,
+        out_shardings=state_sharding,
+    )
+  jitted = bind_mesh(jitted, mesh)
+
+  state = jitted(rng)
+
+  state = max_utils.unbox_logicallypartioned(state)
+  return state
+
+def get_abstract_state_no_sharding(model, tx, config, rng, mesh, is_training=True):
+  """Get a shaped abstraction of the state (including optimizer)"""
+  init_state_partial = functools.partial(init_initial_state, model, tx, config, is_training)
+
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
+    abstract_state = jax.eval_shape(init_state_partial, rng)
+
+  state_logical_annotations = nn.get_partition_spec(abstract_state)
+  with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
+    state_mesh_annotations = nn.logical_to_mesh(state_logical_annotations)
+
+  abstract_state = max_utils.unbox_logicallypartioned(abstract_state)
+  return init_state_partial, abstract_state, state_mesh_annotations
+
 
 def setup_training_state(model, data_iterator, tx, config, rng, mesh, checkpoint_manager):
   is_training = True
@@ -554,6 +711,9 @@ def setup_initial_state(
       model, tx, config, rng, mesh, is_training
   )
 
+  # state is not actually needed right now
+  return None, state_mesh_annotations, state_mesh_shardings, data_iterator
+
   # Initialization
   with nn_partitioning.axis_rules(config.logical_axis_rules):
     restored, raw_params = checkpointing.load_state_if_possible(
@@ -723,7 +883,8 @@ def save_quantized_checkpoint_if_configured(config, params):
 
 def add_config_to_summary_writer(config, summary_writer):
   """Writes config params to tensorboard"""
-  if jax.process_index() == 0:
+  # if jax.process_index() == 0:
+  if False:
     for key, value in config.get_keys().items():
       max_utils.add_text_to_summary_writer(key, str(value), summary_writer)
 
diff --git a/MaxText/pyconfig.py b/MaxText/pyconfig.py
index 60bad660..b2ada115 100644
--- a/MaxText/pyconfig.py
+++ b/MaxText/pyconfig.py
@@ -284,6 +284,7 @@ def validate_model_name(s: str) -> bool:
       "mistral-7b",
       "mixtral-8x7b",
       "mixtral-8x22b",
+      "mixtral-small",
       "deepseek2-16b",
       "deepseek2-236b",
       "deepseek3-671b",
@@ -299,12 +300,22 @@ def validate_model_name(s: str) -> bool:
       "gpt3-22b",
       "gpt3-6b",
       "gpt3-52k",
+      "gpt3-xl",
+      "gpt3-24layers",
       "llama4-17b-16e",
       "llama4-17b-128e",
   )
   if s not in valid_model_names:
     raise ValueError(f"Invalid model name was passed. Got {s}, Valid options {valid_model_names}")
 
+def validate_overrides_consistent(keys1: list[str], keys2: list[str], raw_keys: dict, raw_keys_old: dict):
+  overwritten_keys = [k for k in keys1 if k in keys2]
+  for key in overwritten_keys:
+      if raw_keys_old[key] != raw_keys[key]:
+          raise ValueError(
+              f"Key {key} inconsistently overwritten by argument({raw_keys_old[key]})"
+              f" and model({raw_keys[key]}). This isn't allowed."
+          )
 
 def validate_multimodal_model_name(s: str) -> bool:
   valid_model_names = (
@@ -567,7 +578,7 @@ class _HyperParameters:
     raw_keys["num_slices"] = max_utils.get_num_slices(raw_keys)
     raw_keys["quantization_local_shard_count"] = get_quantization_local_shard_count(raw_keys)
     raw_keys = create_parallelisms_list(raw_keys)
-    raw_keys = set_and_validate_pipeline_config(raw_keys)
+    #raw_keys = set_and_validate_pipeline_config(raw_keys)
 
     if raw_keys["dataset_type"] == "c4_mlperf":
       raw_keys["add_bos"] = False
@@ -654,8 +665,8 @@ class _HyperParameters:
 
 def create_parallelisms_list(raw_keys):
   ici_parallelism = [
-      raw_keys["ici_data_parallelism"],
       raw_keys["ici_pipeline_parallelism"],
+      raw_keys["ici_data_parallelism"],
       raw_keys["ici_fsdp_parallelism"],
       raw_keys["ici_fsdp_transpose_parallelism"],
       raw_keys["ici_sequence_parallelism"],
@@ -668,8 +679,8 @@ def create_parallelisms_list(raw_keys):
       raw_keys["ici_autoregressive_parallelism"],
   ]
   dcn_parallelism = [
-      raw_keys["dcn_data_parallelism"],
       raw_keys["dcn_pipeline_parallelism"],
+      raw_keys["dcn_data_parallelism"],
       raw_keys["dcn_fsdp_parallelism"],
       raw_keys["dcn_fsdp_transpose_parallelism"],
       raw_keys["dcn_sequence_parallelism"],
@@ -898,8 +909,8 @@ def validate_deepseek_moe(raw_keys):
 
 
 def validate_sparse_matmul_parallelism(raw_keys):
-  if raw_keys["sparse_matmul"] and (using_sequence_parallelism(raw_keys) or using_pipeline_parallelism(raw_keys)):
-    raise ValueError("Currently we only support Megablox and Ragged dot with data, tensor, and expert parallelism.")
+  if raw_keys["sparse_matmul"] and (using_sequence_parallelism(raw_keys)):
+    raise ValueError("Currently we only support Megablox and Ragged dot with data, tensor, pipeline, and expert parallelism.")
   tensor_parallelism = (
       raw_keys["ici_tensor_parallelism"]
       * raw_keys["dcn_tensor_parallelism"]
diff --git a/MaxText/train.py b/MaxText/train.py
index 9f06fc37..2c4b9772 100644
--- a/MaxText/train.py
+++ b/MaxText/train.py
@@ -20,7 +20,7 @@ limitations under the License.
 # Calling jax.device_count here prevents a "TPU platform already registered" error.
 # See github.com/google/maxtext/issues/20 for more
 
-from typing import Sequence
+from typing import Optional, Sequence, Union
 import datetime
 import functools
 import os
@@ -74,7 +74,6 @@ from MaxText.metric_logger import MetricLogger
 from MaxText.utils import gcs_utils
 from MaxText.vertex_tensorboard import VertexTensorboardManager
 # Placeholder: internal
-
 # pylint: disable=too-many-positional-arguments
 
 EPS = 1e-8
@@ -130,7 +129,6 @@ def record_scalar_metrics(metrics, step_time_delta, per_device_tflops, lr, per_d
   metrics["scalar"].update({"perf/per_device_tokens_per_sec": per_device_tokens / step_time_delta.total_seconds()})
   metrics["scalar"].update({"learning/current_learning_rate": lr})
 
-
 def save_checkpoint(
     checkpoint_manager,
     step,
@@ -356,7 +354,7 @@ def dpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_t
   return loss, aux
 
 
-def loss_fn(model, config, data, dropout_rng, params, is_train=True):
+def loss_fn(model, config, data, dropout_rng, params, total_weights=None, is_train=True):
   """loss_fn for both train and eval.
 
   Args:
@@ -365,6 +363,7 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
     data: Batch of data to apply to the model
     dropout_rng: A key to use to generate rng for dropout
     params: Model params
+    total_weights: number of non-padding tokens in the entire batch
     is_train: True for train_step and False for eval_step
 
   Returns:
@@ -396,10 +395,13 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
   xent, _ = max_utils.cross_entropy_with_logits(logits, one_hot_targets, 0.0)
   xent = nn.with_logical_constraint(xent, ("activation_embed_and_logits_batch", "activation_length"))
   # Mask out paddings at the end of each example.
-  xent = xent * (data["targets_segmentation"] != 0)
+  weights = data["targets_segmentation"] != 0
+  xent = xent * weights
+  if total_weights is None:
+    total_weights = jnp.sum(weights)
   total_loss = jnp.sum(xent)
-  total_weights = jnp.sum(data["targets_segmentation"] != 0)
-  loss = total_loss / (total_weights + EPS)
+  loss = jnp.sum(xent / (total_weights + EPS))
+
   # get moe load balance loss
   moe_lb_loss = 0.0
   if config.num_experts > 1:
@@ -416,6 +418,22 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
   return loss, aux
 
 
+class MicrobatchConfig:
+  size: Optional[int] = None
+  schedule: str = "wavefront"
+  num_stages: Optional[int] = None
+  interleave: int = 1
+  custom_schedule: Optional[Union[list[list[tuple[int, str]]], list[list[str]]]] = None
+
+microbatch_config = MicrobatchConfig()
+
+def set_mb_config(mb_size, schedule, num_stages, interleave, custom_schedule=None):
+    microbatch_config.size = mb_size
+    microbatch_config.schedule = schedule
+    microbatch_config.num_stages = num_stages
+    microbatch_config.interleave = interleave
+    microbatch_config.custom_schedule = custom_schedule
+
 def train_step(model, config, state_mesh_shardings, state, data, dropout_rng):
   """
 
@@ -478,7 +496,29 @@ def train_step(model, config, state_mesh_shardings, state, data, dropout_rng):
         reference_params = jax.device_put(reference_params, max_utils.with_memory_kind(reference_params_sharding, "device"))
         extra_dpo_args = [reference_params]
     grad_func = jax.value_and_grad(_loss_fn, argnums=4, has_aux=True)
-    (loss, aux), raw_grads = grad_func(model, config, data, dropout_rng, state.params, *extra_dpo_args, is_train=True)
+
+    if microbatch_config.size is None or microbatch_config.size == config.global_batch_size_to_train_on:
+      (loss, aux), raw_grads = grad_func(model, config, data, dropout_rng, state.params, *extra_dpo_args, is_train=True)
+    else:
+      from multimesh.jax import microbatch
+
+      def microbatch_input_sharding(x):
+        return None
+
+      arg_shardings = jax.tree.map(microbatch_input_sharding, data)
+      total_weights = jnp.sum(data["targets_segmentation"] != 0)
+      def mb_grad_func(data, dropout, params):
+        return grad_func(model, config, data, dropout, params, total_weights=total_weights, is_train=True)
+
+      mb_grad_func = microbatch(mb_grad_func, dim=0, argnum=0,
+                             size=microbatch_config.size, arg_shardings=arg_shardings,
+                             schedule=microbatch_config.schedule, num_stages=microbatch_config.num_stages,
+                             interleave=microbatch_config.interleave,
+                             custom_schedule=microbatch_config.custom_schedule)
+
+      (loss, aux), raw_grads = mb_grad_func(data, dropout_rng, state.params)
+
+
   intermediate_outputs = aux["intermediate_outputs"]
   total_weights = aux["total_weights"]
   moe_lb_loss = aux["moe_lb_loss"]
@@ -512,15 +552,19 @@ def train_step(model, config, state_mesh_shardings, state, data, dropout_rng):
 
   scalar_metrics = {
       "learning/loss": loss,
-      "learning/moe_lb_loss": moe_lb_loss,
+      #"learning/moe_lb_loss": moe_lb_loss,
       "learning/total_weights": total_weights,
   }
   if not config.optimizer_memory_host_offload:
-    scalar_metrics["learning/grad_norm"] = max_utils.l2norm_pytree(grads)
-    scalar_metrics["learning/raw_grad_norm"] = max_utils.l2norm_pytree(raw_grads)
-    scalar_metrics["learning/param_norm"] = max_utils.l2norm_pytree(new_state.params)
+    pass
+    #scalar_metrics["learning/grad_norm"] = max_utils.l2norm_pytree(grads)
+    #scalar_metrics["learning/raw_grad_norm"] = max_utils.l2norm_pytree(raw_grads)
+    #scalar_metrics["learning/param_norm"] = max_utils.l2norm_pytree(new_state.params)
+
   if config.use_dpo:
-    scalar_metrics["learning/dpo_reward_accuracy"] = aux["reward_accuracy"]
+    pass
+    #scalar_metrics["learning/dpo_reward_accuracy"] = aux["reward_accuracy"]
+
   metrics = {
       "scalar": scalar_metrics,
       "scalars": {},
@@ -613,8 +657,9 @@ def setup_mesh_and_model(config, devices=None):
   writer = max_utils.initialize_summary_writer(config.tensorboard_dir, config.run_name)
 
   # Mesh definition
+  from multimesh.jax import MultiMesh
   devices_array = maxtext_utils.create_device_mesh(config, devices)
-  mesh = Mesh(devices_array, config.mesh_axes)
+  mesh = MultiMesh(global_mesh=Mesh(devices_array, config.mesh_axes))
 
   # Model and Optimizer definition
   quant = quantizations.configure_quantization(config)
@@ -706,9 +751,9 @@ def setup_train_loop(config):
       model, data_iterator, tx, config, init_rng, mesh, checkpoint_manager
   )
 
-  if not config.using_pipeline_parallelism:
-    # The vocab tensor(s) of shape [vocab, embed] (and transpose) are not sharded by stage
-    maxtext_utils.assert_params_sufficiently_sharded(state.params, mesh, config.sharding_tolerance)
+  #if not config.using_pipeline_parallelism:
+  #  # The vocab tensor(s) of shape [vocab, embed] (and transpose) are not sharded by stage
+  #  maxtext_utils.assert_params_sufficiently_sharded(state.params, mesh, config.sharding_tolerance)
 
   if config.use_dpo:
     abstract_state, _, _ = maxtext_utils.get_abstract_state(model, tx, config, init_rng, mesh, is_training=True)
@@ -738,7 +783,21 @@ def setup_train_loop(config):
       )
 
   record_goodput(recorder, config, recorder.record_training_preparation_end_time if recorder else None)
+
+  # Shaped RNG keys
+  _, example_rng = jax.random.split(jax.random.PRNGKey(0), 2)
+  shaped_rng = jax.ShapeDtypeStruct(example_rng.shape, example_rng.dtype)
+
+  # un-sharded abstract state
+  init_state_partial, abstract_state, state_mesh_annotations = maxtext_utils.get_abstract_state_no_sharding(model, tx, config, example_rng, mesh)
+
+  # Shaped batch
+  shaped_batch = maxtext_utils.get_shaped_batch(config)
+
+  shaped_train_args = (abstract_state, shaped_batch, shaped_rng)
+
   return (
+      init_state_partial,
       init_rng,
       writer,
       checkpoint_manager,
@@ -748,7 +807,7 @@ def setup_train_loop(config):
       learning_rate_schedule,
       data_iterator,
       eval_data_iterator,
-      state,
+      shaped_train_args,
   )
 
 
@@ -765,6 +824,7 @@ def train_loop(config, state=None):
   record_goodput(recorder, config, recorder.record_job_start_time if recorder else None)
 
   (
+      init_state_partial,
       init_rng,
       writer,
       checkpoint_manager,
@@ -774,7 +834,7 @@ def train_loop(config, state=None):
       learning_rate_schedule,
       data_iterator,
       eval_data_iterator,
-      state,
+      shaped_train_args,
   ) = setup_train_loop(config)
 
   if config.use_dpo:
@@ -790,7 +850,8 @@ def train_loop(config, state=None):
       out_shard_train,
       static_argnums_train,
       donate_argnums_train,
-  ) = maxtext_utils.get_functional_train_with_signature(train_step, mesh, state_mesh_shardings, model, config)
+  ) = maxtext_utils.get_and_compile_functional_train_with_signature(train_step, mesh, state_mesh_shardings, model, config, shaped_train_args)
+
 
   if eval_data_iterator:
     # pylint: disable=line-too-long
@@ -800,7 +861,11 @@ def train_loop(config, state=None):
         out_shard_eval,
         static_argnums_eval,
         donate_argnums_eval,
-    ) = maxtext_utils.get_functional_eval_with_signature(eval_step, mesh, state_mesh_shardings, model, config)
+    ) = maxtext_utils.get_functional_eval_with_signature(eval_step, mesh, in_shard_train[0], model, config)
+
+  # postponed initial state fetch as we needed to evaluate sharding first
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
+    state = maxtext_utils.initialize_state_with_sharding(init_state_partial, mesh, in_shard_train[0], in_shard_train[2], init_rng)
 
   num_model_parameters = max_utils.calculate_num_params_from_pytree(state.params)
   max_logging.log(f"number parameters: {num_model_parameters/1e9:.3f} billion")
@@ -821,15 +886,9 @@ def train_loop(config, state=None):
     p_eval_step = None
     print("Loaded compiled function!", flush=True)
   else:
-    p_train_step = jax.jit(
-        functional_train,
-        in_shardings=in_shard_train,
-        out_shardings=out_shard_train,
-        static_argnums=static_argnums_train,
-        donate_argnums=donate_argnums_train,
-    )
+    p_train_step=functional_train
 
-    if eval_data_iterator:
+    if config.eval_interval > 0 and eval_data_iterator:
       p_eval_step = jax.jit(
           functional_eval,
           in_shardings=in_shard_eval,
@@ -873,8 +932,10 @@ def train_loop(config, state=None):
       try:
         example_batch = load_next_batch(data_iterator, example_batch, config)
         # Reshard data from loaded sharding to performant activation sharding
-        example_batch = jax.lax.with_sharding_constraint(example_batch, input_data_shardings)
+        # example_batch = jax.lax.with_sharding_constraint(example_batch, input_data_shardings)
       except Exception as e:  # pylint: disable=broad-except
+        import traceback
+        traceback.print_exc(e)
         max_logging.log(f"load_next_batch failed, you may have run out of data. Error message: {e}")
         break
       record_goodput(recorder, config, recorder.record_data_loading_end_time if recorder else None)
@@ -929,6 +990,7 @@ def train_loop(config, state=None):
       for eval_batch in eval_data_iterator:
         if config.eval_steps > 0 and eval_step_count >= config.eval_steps:
           break
+      
         with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
           eval_metrics = p_eval_step(state, eval_batch, nextrng)
         cumulative_eval_metrics["scalar"]["eval/total_loss"] += float(eval_metrics["scalar"]["evaluation/total_loss"])
@@ -985,7 +1047,11 @@ def train_loop(config, state=None):
   record_goodput(recorder, config, recorder.record_job_end_time if recorder else None)
   with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
     # pytype: disable=attribute-error
-    compiled = p_train_step.lower(state, example_batch, nextrng).compile()
+    if hasattr(p_train_step, "lower"):
+      compiled = p_train_step.lower(state, example_batch, nextrng).compile()
+    else:
+      compiled = p_train_step
+
     compiled_stats = compiled.memory_analysis()
     if compiled_stats is not None:
       max_logging.log(
diff --git a/MaxText/utils/gcs_utils.py b/MaxText/utils/gcs_utils.py
index 83aa2579..f6da2584 100644
--- a/MaxText/utils/gcs_utils.py
+++ b/MaxText/utils/gcs_utils.py
@@ -34,6 +34,7 @@ def write_config_raw_keys_for_gcs(raw_keys):
   """Writes config raw keys to GCS"""
   if not raw_keys["save_config_to_gcs"] or jax.process_index() != 0:
     return
+  return
   max_logging.log("Writing config to GCS...")
 
   raw_keys_dict = dict(raw_keys)
diff --git a/pyproject.toml b/pyproject.toml
index e97cb950..c0755d61 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -29,6 +29,6 @@ dynamic = ["dependencies"]
 dependencies = {file = ["requirements.txt"]}
 
 [tool.setuptools.packages.find]  
-where = ["MaxText"]
+# where = ["MaxText"]
 include = ["MaxText*"]
 exclude = ["MaxText.tests.*"]
diff --git a/requirements.txt b/requirements.txt
index 968a88b4..77d883e0 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,6 +1,6 @@
 jax>=0.4.30
 jaxlib>=0.4.30
-orbax-checkpoint>=0.5.12
+orbax-checkpoint==0.11.15
 absl-py
 array-record
 aqtp
