diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0f86c28fe0..624e2798ea 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -105,6 +105,13 @@ set(SOVERSION 1)

 set(INSTALL_SUFFIX "" CACHE STRING "Suffix to add to installed libraries and binaries")

+set(CMAKE_C_FLAGS_ASAN "${CMAKE_C_FLAGS_RELWITHDEBINFO} -fsanitize=address -fno-omit-frame-pointer -fno-common" CACHE STRING "" FORCE)
+set(CMAKE_CXX_FLAGS_ASAN "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -fsanitize=address -fno-omit-frame-pointer -fno-common" CACHE STRING "" FORCE)
+set(CMAKE_EXE_LINKER_FLAGS_ASAN "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} -fsanitize=address" CACHE STRING "" FORCE)
+set(CMAKE_SHARED_LINKER_FLAGS_ASAN "${CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO} -fsanitize=address" CACHE STRING "" FORCE)
+
+set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo" "ASAN")
+
 #------------------------------------------------------------------------------#
 # C++ standard version
 #------------------------------------------------------------------------------#
diff --git a/runtime/CMakeLists.txt b/runtime/CMakeLists.txt
index 6c1489382a..94c4adb417 100644
--- a/runtime/CMakeLists.txt
+++ b/runtime/CMakeLists.txt
@@ -560,7 +560,6 @@ if(Legion_USE_CUDA)
   target_link_libraries(RealmRuntime PUBLIC CUDA::toolkit)
   target_link_libraries(RealmStatic PUBLIC CUDA::toolkit)
   # Advertise our driver link to Realm consumers while ensuring libcuda.so isn't in Realm's RPATH
-  target_link_libraries(RealmRuntime INTERFACE CUDA::cuda_driver)
   target_link_libraries(RealmStatic INTERFACE CUDA::cuda_driver)
   # for backwards compatibility in applications
   target_compile_definitions(RealmRuntime INTERFACE USE_CUDA)
diff --git a/runtime/realm/inst_layout.inl b/runtime/realm/inst_layout.inl
index 465be72894..0b14ce1c38 100644
--- a/runtime/realm/inst_layout.inl
+++ b/runtime/realm/inst_layout.inl
@@ -113,7 +113,7 @@ namespace Realm {
     InstanceLayout<N,T> *layout = new InstanceLayout<N,T>;
     layout->bytes_used = 0;
     // require 32B alignment of each instance piece for vectorizing goodness
-    layout->alignment_reqd = 32;
+    layout->alignment_reqd = 64;
     layout->space = is;

     // if the index space is empty, all fields can use the same empty
